// @fatsolutions/tongo-sdk/dist/account/account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Account = void 0;
const starknet_1 = require("starknet");
const audit_1 = require("../provers/audit");
const fund_1 = require("../provers/fund");
const ragequit_1 = require("../provers/ragequit");
const rollover_1 = require("../provers/rollover");
const transfer_1 = require("../provers/transfer");
const withdraw_1 = require("../provers/withdraw");
const ae_balance_js_1 = require("../ae_balance.js");
const data_service_js_1 = require("../data.service.js");
const key_js_1 = require("../key.js");
const fund_js_1 = require("../operations/fund.js");
const ragequit_js_1 = require("../operations/ragequit.js");
const rollover_js_1 = require("../operations/rollover.js");
const transfer_js_1 = require("../operations/transfer.js");
const withdraw_js_1 = require("../operations/withdraw.js");
const tongo_abi_js_1 = require("../tongo.abi.js");
const types_js_1 = require("../types.js");
const utils_js_1 = require("../utils.js");
const events_js_1 = require("./events.js");
class Account {
    publicKey;
    pk;
    provider;
    Tongo;
    constructor(pk, contractAddress, providerOrUrl) {
        let provider;
        if (typeof providerOrUrl === "string") {
            provider = new starknet_1.RpcProvider({
                nodeUrl: providerOrUrl,
                specVersion: "0.10.0",
            });
        }
        else {
            provider = providerOrUrl;
        }
        this.pk = (0, utils_js_1.bytesOrNumToBigInt)(pk);
        this.Tongo = new starknet_1.Contract({
            abi: tongo_abi_js_1.tongoAbi,
            address: contractAddress,
            providerOrAccount: provider
        }).typedv2(tongo_abi_js_1.tongoAbi);
        this.publicKey = (0, utils_js_1.pubKeyFromSecret)(this.pk);
        this.provider = provider;
    }
    tongoAddress() {
        return (0, types_js_1.pubKeyAffineToBase58)(this.publicKey);
    }
    static tongoAddress(pk) {
        return (0, types_js_1.pubKeyAffineToBase58)((0, utils_js_1.pubKeyFromSecret)((0, utils_js_1.bytesOrNumToBigInt)(pk)));
    }
    async nonce() {
        const { nonce } = await this.rawState();
        return nonce;
    }
    /// Returns the State of the account. This functions decrypts the balance and pending
    /// CipherBalances.
    async state() {
        const { balance, pending, aeBalance, nonce } = await this.rawState();
        const hint = aeBalance ? await this.decryptAEBalance(aeBalance, nonce) : undefined;
        const balanceAmount = this.decryptCipherBalance(balance, hint);
        const pendingAmount = this.decryptCipherBalance(pending);
        return { balance: balanceAmount, pending: pendingAmount, nonce };
    }
    /// Retunrs the `almost` raw account state. The only handing that happens here is type
    // convertion from CairoOption::None to undefinded and from StarkPoints to ProjectivePoints
    async rawState() {
        const state = await this.Tongo.get_state(this.publicKey);
        return Account.parseAccountState(state);
    }
    /// Returns the rate of convertion of Tongo
    async rate() {
        const rate = await this.Tongo.get_rate();
        return (0, utils_js_1.castBigInt)(rate);
    }
    /// Returns the bit_size of this Tongo contract
    async bit_size() {
        const bit = await this.Tongo.get_bit_size();
        const bit_size = typeof bit == 'bigint' ? Number(bit) : bit;
        return bit_size;
    }
    // Warning: This is only for display. This is not the correct amount
    // of tongos that corresponds to erc20Amount
    async erc20ToTongo(erc20Amount) {
        const rate = await this.rate();
        const temp = erc20Amount / rate;
        if (erc20Amount % rate != 0n) {
            return temp + 1n;
        }
        else {
            return temp;
        }
    }
    async tongoToErc20(tongoAmount) {
        const rate = await this.rate();
        return tongoAmount * rate;
    }
    async auditorKey() {
        const auditorKey = await this.Tongo.auditor_key();
        return auditorKey;
    }
    /// Returns Option(None) if tongo has not and auditor and Some(Audit) if tongo has an auditor
    async createAuditPart(balance, storedCipherBalance, prefix_data) {
        let auditPart = new starknet_1.CairoOption(starknet_1.CairoOptionVariant.None);
        const auditor = await this.auditorKey();
        if (auditor.isSome()) {
            const auditorPubKey = (0, types_js_1.starkPointToProjectivePoint)(auditor.unwrap());
            const { inputs: inputsAudit, proof: proofAudit } = (0, audit_1.proveAudit)(this.pk, balance, storedCipherBalance, auditorPubKey, prefix_data);
            const nonce = await this.nonce();
            const hint = await this.computeAEHintForPubKey(balance, nonce, auditorPubKey);
            const audit = { auditedBalance: inputsAudit.auditedBalance, hint, proof: proofAudit };
            auditPart = new starknet_1.CairoOption(starknet_1.CairoOptionVariant.Some, audit);
        }
        return auditPart;
    }
    async fund(fundDetails) {
        const { amount, sender } = fundDetails;
        const { nonce, balance: currentBalance, aeBalance } = await this.rawState();
        const current_hint = aeBalance ? await this.decryptAEBalance(aeBalance, nonce) : undefined;
        const initialBalance = this.decryptCipherBalance(currentBalance, current_hint);
        const prefix_data = {
            chain_id: BigInt(await this.provider.getChainId()),
            tongo_address: BigInt(this.Tongo.address),
            sender_address: BigInt(sender)
        };
        const { inputs, proof, newBalance } = (0, fund_1.proveFund)(this.pk, amount, initialBalance, currentBalance, nonce, prefix_data);
        //audit
        const auditPart = await this.createAuditPart(amount + initialBalance, newBalance, prefix_data);
        const hint = await this.computeAEHintForSelf(amount + initialBalance, nonce + 1n);
        const operation = new fund_js_1.FundOperation({ to: inputs.y, amount, hint, proof, auditPart, Tongo: this.Tongo });
        await operation.populateApprove();
        return operation;
    }
    async transfer(transferDetails) {
        const { amount, sender } = transferDetails;
        const bit_size = await this.bit_size();
        const { nonce, balance: currentBalance, aeBalance } = await this.rawState();
        const current_hint = aeBalance ? await this.decryptAEBalance(aeBalance, nonce) : undefined;
        const initialBalance = this.decryptCipherBalance(currentBalance, current_hint);
        if (initialBalance < amount) {
            throw new Error("You dont have enough balance");
        }
        const to = (0, types_js_1.starkPointToProjectivePoint)(transferDetails.to);
        const prefix_data = {
            chain_id: BigInt(await this.provider.getChainId()),
            tongo_address: BigInt(this.Tongo.address),
            sender_address: BigInt(sender),
        };
        const { inputs, proof, newBalance } = (0, transfer_1.proveTransfer)(this.pk, to, initialBalance, amount, currentBalance, nonce, bit_size, prefix_data);
        const hintTransfer = await this.computeAEHintForPubKey(amount, nonce, to);
        const hintLeftover = await this.computeAEHintForSelf(initialBalance - amount, nonce + 1n);
        //audit
        const auditPart = await this.createAuditPart(initialBalance - amount, newBalance, prefix_data);
        const auditPartTransfer = await this.createAuditPart(amount, inputs.transferBalanceSelf, prefix_data);
        return new transfer_js_1.TransferOperation({
            from: inputs.from,
            to: inputs.to,
            transferBalance: inputs.transferBalance,
            transferBalanceSelf: inputs.transferBalanceSelf,
            auxiliarCipher: inputs.auxiliarCipher,
            auxiliarCipher2: inputs.auxiliarCipher2,
            hintTransfer,
            hintLeftover,
            proof,
            auditPart,
            auditPartTransfer,
            Tongo: this.Tongo,
        });
    }
    async ragequit(ragequitDetails) {
        const { to, sender } = ragequitDetails;
        const { nonce, balance: currentBalance, aeBalance } = await this.rawState();
        const current_hint = aeBalance ? await this.decryptAEBalance(aeBalance, nonce) : undefined;
        const currentBalanceAmount = this.decryptCipherBalance(currentBalance, current_hint);
        if (currentBalanceAmount === 0n) {
            throw new Error("You dont have enough balance");
        }
        const prefix_data = {
            chain_id: BigInt(await this.provider.getChainId()),
            tongo_address: BigInt(this.Tongo.address),
            sender_address: BigInt(sender),
        };
        const { inputs, proof, newBalance } = (0, ragequit_1.proveRagequit)(this.pk, currentBalance, nonce, BigInt(to), currentBalanceAmount, prefix_data);
        // zeroing out aehints
        const hint = await this.computeAEHintForSelf(0n, nonce + 1n);
        const auditPart = await this.createAuditPart(0n, newBalance, prefix_data);
        return new ragequit_js_1.RagequitOperation({
            from: inputs.y,
            to: inputs.to,
            amount: inputs.amount,
            hint,
            proof,
            Tongo: this.Tongo,
            auditPart,
        });
    }
    async withdraw(withdrawDetails) {
        const { amount, to, sender } = withdrawDetails;
        const bit_size = await this.bit_size();
        const { nonce, balance: currentBalance, aeBalance } = await this.rawState();
        const current_hint = aeBalance ? await this.decryptAEBalance(aeBalance, nonce) : undefined;
        const initialBalance = this.decryptCipherBalance(currentBalance, current_hint);
        if (initialBalance < amount) {
            throw new Error("You dont have enought balance");
        }
        const prefix_data = {
            chain_id: BigInt(await this.provider.getChainId()),
            tongo_address: BigInt(this.Tongo.address),
            sender_address: BigInt(sender),
        };
        const { inputs, proof, newBalance } = (0, withdraw_1.proveWithdraw)(this.pk, initialBalance, amount, BigInt(to), currentBalance, nonce, bit_size, prefix_data);
        const hint = await this.computeAEHintForSelf(initialBalance - amount, nonce + 1n);
        //audit
        const auditPart = await this.createAuditPart(initialBalance - amount, newBalance, prefix_data);
        return new withdraw_js_1.WithdrawOperation({
            from: inputs.y,
            to: inputs.to,
            amount: inputs.amount,
            auxiliarCipher: inputs.auxiliarCipher,
            hint,
            proof,
            auditPart,
            Tongo: this.Tongo,
        });
    }
    async rollover(rolloverDetails) {
        const { sender } = rolloverDetails;
        const state = await this.rawState();
        const { nonce, balance: currentBalance, aeBalance, pending } = state;
        const current_hint = aeBalance ? await this.decryptAEBalance(aeBalance, nonce) : undefined;
        const unlockedAmount = this.decryptCipherBalance(currentBalance, current_hint);
        const pendingAmount = this.decryptCipherBalance(pending);
        if (pendingAmount == 0n) {
            throw new Error("Your pending ammount is 0");
        }
        const prefix_data = {
            chain_id: BigInt(await this.provider.getChainId()),
            tongo_address: BigInt(this.Tongo.address),
            sender_address: BigInt(sender),
        };
        const { inputs, proof } = (0, rollover_1.proveRollover)(this.pk, nonce, prefix_data);
        const hint = await this.computeAEHintForSelf(pendingAmount + unlockedAmount, nonce + 1n);
        return new rollover_js_1.RollOverOperation({ to: inputs.y, proof, Tongo: this.Tongo, hint });
    }
    async decryptAEBalance(aeBalance, accountNonce) {
        return this.decryptAEHintForPubKey(aeBalance, accountNonce, this.publicKey);
    }
    async decryptAEHintForPubKey(aeHint, accountNonce, other) {
        const keyAEHint = await this.deriveSymmetricKeyForPubKey(accountNonce, other);
        const { ciphertext, nonce: cipherNonce } = (0, ae_balance_js_1.AEHintToBytes)(aeHint);
        const balance = new ae_balance_js_1.AEChaCha(keyAEHint).decryptBalance({ ciphertext, nonce: cipherNonce });
        return balance;
    }
    decryptCipherBalance({ L, R }, hint) {
        if (hint) {
            if ((0, utils_js_1.assertBalance)(this.pk, hint, L, R)) {
                return hint;
            }
        }
        return (0, utils_js_1.decipherBalance)(this.pk, L, R);
    }
    //TODO: rethink this to better ux
    async generateExPost(to, cipher, sender) {
        if (cipher.L == null) {
            throw new Error("L is null");
        }
        if (cipher.R == null) {
            throw new Error("R is null");
        }
        const prefix_data = {
            chain_id: BigInt(await this.provider.getChainId()),
            tongo_address: BigInt(this.Tongo.address),
            sender_address: BigInt(sender),
        };
        const balance = this.decryptCipherBalance(cipher);
        const { inputs, proof } = (0, audit_1.proveAudit)(this.pk, balance, cipher, (0, types_js_1.starkPointToProjectivePoint)(to), prefix_data);
        return { inputs, proof };
    }
    verifyExPost(expost) {
        const y = (0, types_js_1.projectivePointToStarkPoint)(expost.inputs.y);
        if (y != this.publicKey) {
            throw new Error("The expost is not for you");
        }
        (0, audit_1.verifyAudit)(expost.inputs, expost.proof);
        const amount = this.decryptCipherBalance({
            L: expost.inputs.auditedBalance.L,
            R: expost.inputs.auditedBalance.R,
        });
        return amount;
    }
    _diffieHellman(other) {
        const otherPublicKey = (0, types_js_1.pubKeyBase58ToHex)(other);
        return (0, key_js_1.ECDiffieHellman)(this.pk, otherPublicKey);
    }
    async computeAEHintForPubKey(amount, nonce, pubKey) {
        const keyAEBal = await this.deriveSymmetricKeyForPubKey(nonce, pubKey);
        return (0, ae_balance_js_1.bytesToAEHint)(new ae_balance_js_1.AEChaCha(keyAEBal).encryptBalance(amount));
    }
    async computeAEHintForSelf(amount, nonce) {
        return this.computeAEHintForPubKey(amount, nonce, this.publicKey);
    }
    async deriveSymmetricKeyForPubKey(nonce, other) {
        const sharedSecret = (0, key_js_1.ECDiffieHellman)(this.pk, (0, types_js_1.pubKeyAffineToHex)(other));
        return (0, key_js_1.deriveSymmetricEncryptionKey)({
            contractAddress: this.Tongo.address,
            nonce,
            secret: sharedSecret,
        });
    }
    static parseAccountState(state) {
        const { balance, pending, audit, nonce, ae_balance, ae_audit_balance } = state;
        let parsedAudit;
        if (audit.isSome()) {
            parsedAudit = (0, types_js_1.parseCipherBalance)(audit.unwrap());
        }
        return {
            balance: (0, types_js_1.parseCipherBalance)(balance),
            pending: (0, types_js_1.parseCipherBalance)(pending),
            audit: parsedAudit,
            nonce: starknet_1.num.toBigInt(nonce),
            aeBalance: ae_balance.isSome() ? (0, ae_balance_js_1.parseAEBalance)(ae_balance.unwrap()) : undefined,
            aeAuditBalance: ae_audit_balance.isSome() ? (0, ae_balance_js_1.parseAEBalance)(ae_audit_balance.unwrap()) : undefined,
        };
    }
    async getEventsFund(initialBlock) {
        const reader = new data_service_js_1.StarknetEventReader(this.provider, this.Tongo.address);
        const events = await reader.getEventsFund(initialBlock, this.publicKey);
        return events.map((event) => ({
            type: events_js_1.ReaderToAccountEvents[event.type],
            tx_hash: event.tx_hash,
            block_number: event.block_number,
            nonce: event.nonce,
            amount: event.amount,
        }));
    }
    async getEventsRollover(initialBlock) {
        const reader = new data_service_js_1.StarknetEventReader(this.provider, this.Tongo.address);
        const events = await reader.getEventsRollover(initialBlock, this.publicKey);
        return events.map((event) => ({
            type: events_js_1.ReaderToAccountEvents[event.type],
            tx_hash: event.tx_hash,
            block_number: event.block_number,
            nonce: event.nonce,
            amount: this.decryptCipherBalance((0, types_js_1.parseCipherBalance)(event.rollovered)),
        }));
    }
    async getEventsWithdraw(initialBlock) {
        const reader = new data_service_js_1.StarknetEventReader(this.provider, this.Tongo.address);
        const events = await reader.getEventsWithdraw(initialBlock, this.publicKey);
        return events.map((event) => ({
            type: events_js_1.ReaderToAccountEvents[event.type],
            tx_hash: event.tx_hash,
            block_number: event.block_number,
            nonce: event.nonce,
            amount: event.amount,
            to: starknet_1.num.toHex(event.to),
        }));
    }
    async getEventsRagequit(initialBlock) {
        const reader = new data_service_js_1.StarknetEventReader(this.provider, this.Tongo.address);
        const events = await reader.getEventsRagequit(initialBlock, this.publicKey);
        return events.map((event) => ({
            type: events_js_1.ReaderToAccountEvents[event.type],
            tx_hash: event.tx_hash,
            block_number: event.block_number,
            nonce: event.nonce,
            amount: event.amount,
            to: starknet_1.num.toHex(event.to),
        }));
    }
    async getEventsTransferOut(initialBlock) {
        const reader = new data_service_js_1.StarknetEventReader(this.provider, this.Tongo.address);
        const events = await reader.getEventsTransferOut(initialBlock, this.publicKey);
        return Promise.all(events.map(async (event) => ({
            type: events_js_1.ReaderToAccountEvents[event.type],
            tx_hash: event.tx_hash,
            block_number: event.block_number,
            nonce: event.nonce,
            amount: this.decryptCipherBalance((0, types_js_1.parseCipherBalance)(event.transferBalanceSelf), await this.decryptAEHintForPubKey(event.hintTransfer, event.nonce, event.to)),
            to: (0, types_js_1.pubKeyAffineToBase58)(event.to),
        })));
    }
    async getEventsTransferIn(initialBlock) {
        const reader = new data_service_js_1.StarknetEventReader(this.provider, this.Tongo.address);
        const events = await reader.getEventsTransferIn(initialBlock, this.publicKey);
        return Promise.all(events.map(async (event) => ({
            type: events_js_1.ReaderToAccountEvents[event.type],
            tx_hash: event.tx_hash,
            block_number: event.block_number,
            nonce: event.nonce,
            amount: this.decryptCipherBalance((0, types_js_1.parseCipherBalance)(event.transferBalance), await this.decryptAEHintForPubKey(event.hintTransfer, event.nonce, event.from)),
            from: (0, types_js_1.pubKeyAffineToBase58)(event.from),
        })));
    }
    async getTxHistory(initialBlock) {
        const promises = Promise.all([
            this.getEventsFund(initialBlock),
            this.getEventsRollover(initialBlock),
            this.getEventsWithdraw(initialBlock),
            this.getEventsRagequit(initialBlock),
            this.getEventsTransferOut(initialBlock),
            this.getEventsTransferIn(initialBlock),
        ]);
        const events = (await promises).flat();
        return events.sort((a, b) => b.block_number - a.block_number);
    }
}
exports.Account = Account;
