// @fatsolutions/tongo-sdk/dist/ae_balance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AEChaCha = void 0;
exports.AEHintToBytes = AEHintToBytes;
exports.bytesToAEHint = bytesToAEHint;
exports.parseAEBalance = parseAEBalance;
const chacha_js_1 = require("@noble/ciphers/chacha.js");
const utils_js_1 = require("@noble/ciphers/utils.js");
const webcrypto_js_1 = require("@noble/ciphers/webcrypto.js");
const starknet_1 = require("starknet");
const utils_js_2 = require("./utils.js");
function AEHintToBytes({ ciphertext, nonce }) {
    return {
        ciphertext: (0, utils_js_1.numberToBytesBE)(ciphertext, 64),
        nonce: (0, utils_js_1.numberToBytesBE)(nonce, 24), // XChaCha20 nonce is 192 bits
    };
}
function bytesToAEHint({ ciphertext, nonce }) {
    return {
        ciphertext: (0, utils_js_1.bytesToNumberBE)(ciphertext),
        nonce: (0, utils_js_1.bytesToNumberBE)(nonce),
    };
}
function parseAEBalance({ ciphertext, nonce, }) {
    let parsedNonce;
    if ((0, utils_js_2.isUint256)(nonce)) {
        parsedNonce = starknet_1.uint256.uint256ToBN(nonce);
    }
    else {
        parsedNonce = BigInt(nonce);
    }
    return {
        ciphertext: BigInt(ciphertext),
        nonce: parsedNonce,
    };
}
// TODO: we should split this class into a AECipher class that returns
// AEBalance's. This way we can decouple the cipher from the balance and its
// serialization.
class AEChaCha {
    key;
    constructor(key) {
        this.key = key;
        // 32 B
        if (this.key.length != 32) {
            throw new Error(`Key length must be exactly 32 Bytes, not '${this.key.length}'`);
        }
    }
    encryptBalance(balance) {
        // TODO: refactor AEChaCha to support different bit sizes. (Max is 384 bits)
        if (balance >= 2n ** 32n) {
            throw new Error("This implementation only supports 32 bit balances");
        }
        // 512  = ( TAG [128] ) + ( NOISE/RESERVED [352] ) + ( BALANCE [32] )
        // 64 B      16 B                44 B                    4 B
        const nonce = (0, webcrypto_js_1.randomBytes)(24); // XChaCha20 uses random nonces of 192 bit = 24 B
        const noise = (0, webcrypto_js_1.randomBytes)(3 * 16 - 4);
        const numberBytes = (0, utils_js_1.numberToBytesBE)(balance, 48);
        numberBytes.set(noise, 0);
        const chacha = (0, chacha_js_1.xchacha20poly1305)(this.key, nonce);
        const ciphertext = chacha.encrypt(numberBytes);
        return { ciphertext, nonce };
    }
    decryptBalance({ ciphertext, nonce }) {
        const chacha = (0, chacha_js_1.xchacha20poly1305)(this.key, nonce);
        try {
            const plaintext = chacha.decrypt(ciphertext);
            if (plaintext.length !== 48)
                throw new Error("Malformed plaintext");
            return (0, utils_js_1.bytesToNumberBE)(plaintext.slice(44, 48));
        }
        catch {
            throw new Error("Malformed or tampered ciphertext");
        }
    }
}
exports.AEChaCha = AEChaCha;
