// @fatsolutions/tongo-sdk/dist/auditor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Auditor = exports.AuditorEvent = void 0;
const starknet_1 = require("starknet");
const data_service_js_1 = require("./data.service.js");
const utils_1 = require("./utils");
const types_js_1 = require("./types.js");
const ae_balance_js_1 = require("./ae_balance.js");
const key_js_1 = require("./key.js");
const tongo_abi_js_1 = require("./tongo.abi.js");
const utils_js_1 = require("./utils.js");
exports.AuditorEvent = {
    BalanceDeclared: 'balanceDeclared',
    TransferOutDeclared: 'transferOutDeclared',
    TransferInDeclared: 'transferInDeclared',
};
class Auditor {
    publicKey;
    pk;
    provider;
    Tongo;
    constructor(pk, contractAddress, provider) {
        this.pk = (0, utils_js_1.bytesOrNumToBigInt)(pk);
        this.Tongo = new starknet_1.Contract({
            abi: tongo_abi_js_1.tongoAbi,
            address: contractAddress,
            providerOrAccount: provider
        }).typedv2(tongo_abi_js_1.tongoAbi);
        this.publicKey = (0, types_js_1.derivePublicKey)(this.pk);
        this.provider = provider;
    }
    decryptCipherBalance({ L, R }, hint) {
        if (hint) {
            if ((0, utils_1.assertBalance)(this.pk, hint, L, R)) {
                return hint;
            }
        }
        return (0, utils_1.decipherBalance)(this.pk, L, R);
    }
    async deriveSymmetricKeyForPubKey(nonce, other) {
        const sharedSecret = (0, key_js_1.ECDiffieHellman)(this.pk, (0, types_js_1.pubKeyAffineToHex)(other));
        return (0, key_js_1.deriveSymmetricEncryptionKey)({
            contractAddress: this.Tongo.address,
            nonce,
            secret: sharedSecret,
        });
    }
    async decryptAEHintForPubKey(aeHint, accountNonce, other) {
        const keyAEHint = await this.deriveSymmetricKeyForPubKey(accountNonce, other);
        const { ciphertext, nonce: cipherNonce } = (0, ae_balance_js_1.AEHintToBytes)(aeHint);
        const balance = new ae_balance_js_1.AEChaCha(keyAEHint).decryptBalance({ ciphertext, nonce: cipherNonce });
        return balance;
    }
    async getUserBalance(initialBlock, otherPubKey) {
        const reader = new data_service_js_1.StarknetEventReader(this.provider, this.Tongo.address);
        const events = await reader.getEventsBalanceDeclared(initialBlock, otherPubKey);
        return Promise.all(events.map(async (event) => ({
            type: exports.AuditorEvent.BalanceDeclared,
            tx_hash: event.tx_hash,
            block_number: event.block_number,
            nonce: event.nonce,
            user: (0, types_js_1.pubKeyAffineToBase58)(otherPubKey),
            amount: this.decryptCipherBalance((0, types_js_1.parseCipherBalance)(event.declaredCipherBalance), await this.decryptAEHintForPubKey(event.hint, event.nonce, otherPubKey)),
        })));
    }
    async getUserTransferOut(initialBlock, otherPubKey) {
        const reader = new data_service_js_1.StarknetEventReader(this.provider, this.Tongo.address);
        const events = await reader.getEventsTransferFrom(initialBlock, otherPubKey);
        return Promise.all(events.map(async (event) => ({
            type: exports.AuditorEvent.TransferOutDeclared,
            tx_hash: event.tx_hash,
            block_number: event.block_number,
            sender_nonce: event.nonce,
            user: (0, types_js_1.pubKeyAffineToBase58)(otherPubKey),
            amount: this.decryptCipherBalance((0, types_js_1.parseCipherBalance)(event.declaredCipherBalance), await this.decryptAEHintForPubKey(event.hint, event.nonce, otherPubKey)),
            to: (0, types_js_1.pubKeyAffineToBase58)(event.to),
        })));
    }
    async getUserTransferIn(initialBlock, otherPubKey) {
        const reader = new data_service_js_1.StarknetEventReader(this.provider, this.Tongo.address);
        const events = await reader.getEventsTransferTo(initialBlock, otherPubKey);
        return Promise.all(events.map(async (event) => ({
            type: exports.AuditorEvent.TransferInDeclared,
            tx_hash: event.tx_hash,
            block_number: event.block_number,
            sender_nonce: event.nonce,
            user: (0, types_js_1.pubKeyAffineToBase58)(otherPubKey),
            amount: this.decryptCipherBalance((0, types_js_1.parseCipherBalance)(event.declaredCipherBalance), await this.decryptAEHintForPubKey(event.hint, event.nonce, event.from)),
            from: (0, types_js_1.pubKeyAffineToBase58)(event.from),
        })));
    }
    async getUserHistory(initialBlock, user) {
        const promises = Promise.all([
            this.getUserBalance(initialBlock, user),
            this.getUserTransferOut(initialBlock, user),
            this.getUserTransferIn(initialBlock, user),
        ]);
        const events = (await promises).flat();
        return events.sort((a, b) => b.block_number - a.block_number);
    }
}
exports.Auditor = Auditor;
