// @fatsolutions/tongo-sdk/dist/data.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StarknetEventReader = exports.ReaderEventType = void 0;
const starknet_1 = require("starknet");
const tongo_abi_js_1 = require("./tongo.abi.js");
const abiEvents = starknet_1.events.getAbiEvents(tongo_abi_js_1.tongoAbi);
const abiStructs = starknet_1.CallData.getAbiStruct(tongo_abi_js_1.tongoAbi);
const abiEnums = starknet_1.CallData.getAbiEnum(tongo_abi_js_1.tongoAbi);
const FUND_EVENT = starknet_1.num.toHex(starknet_1.hash.starknetKeccak("FundEvent"));
const ROLLOVER_EVENT = starknet_1.num.toHex(starknet_1.hash.starknetKeccak("RolloverEvent"));
const TRANSFER_EVENT = starknet_1.num.toHex(starknet_1.hash.starknetKeccak("TransferEvent"));
const WITHDRAW_EVENT = starknet_1.num.toHex(starknet_1.hash.starknetKeccak("WithdrawEvent"));
const RAGEQUIT_EVENT = starknet_1.num.toHex(starknet_1.hash.starknetKeccak("RagequitEvent"));
const BALANCE_DECLARED_EVENT = starknet_1.num.toHex(starknet_1.hash.starknetKeccak("BalanceDeclared"));
const TRANSFER_DECLARED_EVENT = starknet_1.num.toHex(starknet_1.hash.starknetKeccak("TransferDeclared"));
const FUND_EVENT_PATH = "tongo::structs::events::FundEvent";
const ROLLOVER_EVENT_PATH = "tongo::structs::events::RolloverEvent";
const TRANSFER_EVENT_PATH = "tongo::structs::events::TransferEvent";
const WITHDRAW_EVENT_PATH = "tongo::structs::events::WithdrawEvent";
const RAGEQUIT_EVENT_PATH = "tongo::structs::events::RagequitEvent";
const BALANCE_DECLARED_EVENT_PATH = "tongo::structs::events::BalanceDeclared";
const TRANSFER_DECLARED_EVENT_PATH = "tongo::structs::events::TransferDeclared";
exports.ReaderEventType = {
    Fund: "fund",
    Withdraw: "withdraw",
    Ragequit: "ragequit",
    Rollover: "rollover",
    TransferIn: "transferIn",
    TransferOut: "transferOut",
    BalanceDeclared: "balanceDeclared",
    TransferDeclared: "transferDeclared",
};
function parseTransferEventOut(event) {
    const data = event[TRANSFER_EVENT_PATH];
    return {
        type: exports.ReaderEventType.TransferOut,
        tx_hash: event.transaction_hash,
        block_number: event.block_number,
        ...data,
    };
}
function parseTransferEventIn(event) {
    const data = event[TRANSFER_EVENT_PATH];
    return {
        type: exports.ReaderEventType.TransferIn,
        tx_hash: event.transaction_hash,
        block_number: event.block_number,
        ...data,
    };
}
function parseFundEvent(event) {
    const data = event[FUND_EVENT_PATH];
    return {
        type: exports.ReaderEventType.Fund,
        tx_hash: event.transaction_hash,
        block_number: event.block_number,
        ...data,
    };
}
function parseWithdrawEvent(event) {
    const data = event[WITHDRAW_EVENT_PATH];
    return {
        type: exports.ReaderEventType.Withdraw,
        tx_hash: event.transaction_hash,
        block_number: event.block_number,
        ...data,
    };
}
function parseRagequitEvent(event) {
    const data = event[RAGEQUIT_EVENT_PATH];
    return {
        type: exports.ReaderEventType.Ragequit,
        tx_hash: event.transaction_hash,
        block_number: event.block_number,
        ...data,
    };
}
function parseRolloverEvent(event) {
    const data = event[ROLLOVER_EVENT_PATH];
    return {
        type: exports.ReaderEventType.Rollover,
        tx_hash: event.transaction_hash,
        block_number: event.block_number,
        ...data,
    };
}
function parseBalanceDeclaredEvent(event) {
    const data = event[BALANCE_DECLARED_EVENT_PATH];
    return {
        type: exports.ReaderEventType.BalanceDeclared,
        tx_hash: event.transaction_hash,
        block_number: event.block_number,
        ...data,
    };
}
function parseTransferDeclaredEvent(event) {
    const data = event[TRANSFER_DECLARED_EVENT_PATH];
    return {
        type: exports.ReaderEventType.TransferDeclared,
        tx_hash: event.transaction_hash,
        block_number: event.block_number,
        ...data,
    };
}
class StarknetEventReader {
    provider;
    static abiParser = new starknet_1.AbiParser2(tongo_abi_js_1.tongoAbi);
    abiParser = StarknetEventReader.abiParser;
    tongoAddress;
    constructor(provider, tongoAddress) {
        this.provider = provider;
        this.tongoAddress = tongoAddress;
    }
    async getEventsFund(initialBlock, otherPubKey) {
        const eventsResults = await this.provider.getEvents({
            address: this.tongoAddress,
            from_block: { block_number: initialBlock },
            to_block: "latest",
            keys: [[FUND_EVENT], [starknet_1.num.toHex(otherPubKey.x)], [starknet_1.num.toHex(otherPubKey.y)]],
            chunk_size: 100,
        });
        const parsedEvents = starknet_1.events.parseEvents(eventsResults.events, abiEvents, abiStructs, abiEnums, this.abiParser);
        return parsedEvents
            .filter((event) => event[FUND_EVENT_PATH] !== undefined)
            .map((event) => parseFundEvent(event));
    }
    async getEventsWithdraw(initialBlock, otherPubKey) {
        const eventsResults = await this.provider.getEvents({
            address: this.tongoAddress,
            from_block: { block_number: initialBlock },
            to_block: "latest",
            keys: [[WITHDRAW_EVENT], [starknet_1.num.toHex(otherPubKey.x)], [starknet_1.num.toHex(otherPubKey.y)]],
            chunk_size: 100,
        });
        const parsedEvents = starknet_1.events.parseEvents(eventsResults.events, abiEvents, abiStructs, abiEnums, this.abiParser);
        return parsedEvents
            .filter((event) => event[WITHDRAW_EVENT_PATH] !== undefined)
            .map((event) => parseWithdrawEvent(event));
    }
    async getEventsRagequit(initialBlock, otherPubKey) {
        const eventsResults = await this.provider.getEvents({
            address: this.tongoAddress,
            from_block: { block_number: initialBlock },
            to_block: "latest",
            keys: [[RAGEQUIT_EVENT], [starknet_1.num.toHex(otherPubKey.x)], [starknet_1.num.toHex(otherPubKey.y)]],
            chunk_size: 100,
        });
        const parsedEvents = starknet_1.events.parseEvents(eventsResults.events, abiEvents, abiStructs, abiEnums, this.abiParser);
        return parsedEvents
            .filter((event) => event[RAGEQUIT_EVENT_PATH] !== undefined)
            .map((event) => parseRagequitEvent(event));
    }
    async getEventsRollover(initialBlock, otherPubKey) {
        const eventsResults = await this.provider.getEvents({
            address: this.tongoAddress,
            from_block: { block_number: initialBlock },
            to_block: "latest",
            keys: [[ROLLOVER_EVENT], [starknet_1.num.toHex(otherPubKey.x)], [starknet_1.num.toHex(otherPubKey.y)]],
            chunk_size: 100,
        });
        const parsedEvents = starknet_1.events.parseEvents(eventsResults.events, abiEvents, abiStructs, abiEnums, this.abiParser);
        return parsedEvents
            .filter((event) => event[ROLLOVER_EVENT_PATH] !== undefined)
            .map((event) => parseRolloverEvent(event));
    }
    async getEventsTransferOut(initialBlock, otherPubKey) {
        const eventsResults = await this.provider.getEvents({
            address: this.tongoAddress,
            from_block: { block_number: initialBlock },
            to_block: "latest",
            keys: [[TRANSFER_EVENT], [], [], [starknet_1.num.toHex(otherPubKey.x)], [starknet_1.num.toHex(otherPubKey.y)], []],
            chunk_size: 100,
        });
        const parsedEvents = starknet_1.events.parseEvents(eventsResults.events, abiEvents, abiStructs, abiEnums, this.abiParser);
        return parsedEvents
            .filter((event) => event[TRANSFER_EVENT_PATH] !== undefined)
            .map((event) => parseTransferEventOut(event));
    }
    async getEventsTransferIn(initialBlock, otherPubKey) {
        const eventsResults = await this.provider.getEvents({
            address: this.tongoAddress,
            from_block: { block_number: initialBlock },
            to_block: "latest",
            keys: [[TRANSFER_EVENT], [starknet_1.num.toHex(otherPubKey.x)], [starknet_1.num.toHex(otherPubKey.y)], [], [], []],
            chunk_size: 100,
        });
        const parsedEvents = starknet_1.events.parseEvents(eventsResults.events, abiEvents, abiStructs, abiEnums, this.abiParser);
        return parsedEvents
            .filter((event) => event[TRANSFER_EVENT_PATH] !== undefined)
            .map((event) => parseTransferEventIn(event));
    }
    async getAllEvents(initialBlock, otherPubKey) {
        const promises = Promise.all([
            this.getEventsFund(initialBlock, otherPubKey),
            this.getEventsRollover(initialBlock, otherPubKey),
            this.getEventsWithdraw(initialBlock, otherPubKey),
            this.getEventsRagequit(initialBlock, otherPubKey),
            this.getEventsTransferOut(initialBlock, otherPubKey),
            this.getEventsTransferIn(initialBlock, otherPubKey),
        ]);
        const events = (await promises).flat();
        return events.sort((a, b) => b.block_number - a.block_number);
    }
    async getEventsBalanceDeclared(initialBlock, otherPubKey) {
        const eventsResults = await this.provider.getEvents({
            address: this.tongoAddress,
            from_block: { block_number: initialBlock },
            to_block: "latest",
            keys: [[BALANCE_DECLARED_EVENT], [starknet_1.num.toHex(otherPubKey.x)], [starknet_1.num.toHex(otherPubKey.y)], [], [], []],
            chunk_size: 100,
        });
        const parsedEvents = starknet_1.events.parseEvents(eventsResults.events, abiEvents, abiStructs, abiEnums, this.abiParser);
        return parsedEvents
            .filter((event) => event[BALANCE_DECLARED_EVENT_PATH] !== undefined)
            .map((event) => parseBalanceDeclaredEvent(event));
    }
    async getEventsTransferFrom(initialBlock, otherPubKey) {
        const eventsResults = await this.provider.getEvents({
            address: this.tongoAddress,
            from_block: { block_number: initialBlock },
            to_block: "latest",
            keys: [[TRANSFER_DECLARED_EVENT], [starknet_1.num.toHex(otherPubKey.x)], [starknet_1.num.toHex(otherPubKey.y)], [], [], []],
            chunk_size: 100,
        });
        const parsedEvents = starknet_1.events.parseEvents(eventsResults.events, abiEvents, abiStructs, abiEnums, this.abiParser);
        return parsedEvents
            .filter((event) => event[TRANSFER_DECLARED_EVENT_PATH] !== undefined)
            .map((event) => parseTransferDeclaredEvent(event));
    }
    async getEventsTransferTo(initialBlock, otherPubKey) {
        const eventsResults = await this.provider.getEvents({
            address: this.tongoAddress,
            from_block: { block_number: initialBlock },
            to_block: "latest",
            keys: [[TRANSFER_DECLARED_EVENT], [], [], [starknet_1.num.toHex(otherPubKey.x)], [starknet_1.num.toHex(otherPubKey.y)], []],
            chunk_size: 100,
        });
        const parsedEvents = starknet_1.events.parseEvents(eventsResults.events, abiEvents, abiStructs, abiEnums, this.abiParser);
        return parsedEvents
            .filter((event) => event[TRANSFER_DECLARED_EVENT_PATH] !== undefined)
            .map((event) => parseTransferDeclaredEvent(event));
    }
}
exports.StarknetEventReader = StarknetEventReader;
