// @fatsolutions/tongo-sdk/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectivePoint = void 0;
exports.createCipherBalance = createCipherBalance;
exports.compute_prefix = compute_prefix;
exports.starkPointToProjectivePoint = starkPointToProjectivePoint;
exports.projectivePointToStarkPoint = projectivePointToStarkPoint;
exports.derivePublicKey = derivePublicKey;
exports.pubKeyAffineToHex = pubKeyAffineToHex;
exports.pubKeyAffineToBase58 = pubKeyAffineToBase58;
exports.pubKeyBase58ToAffine = pubKeyBase58ToAffine;
exports.pubKeyBase58ToHex = pubKeyBase58ToHex;
exports.parseCipherBalance = parseCipherBalance;
const she_1 = require("@fatsolutions/she");
const base_1 = require("@scure/base");
const utils_1 = require("@noble/hashes/utils");
const starknet_1 = require("@scure/starknet");
const constants_1 = require("./constants");
exports.ProjectivePoint = she_1.ProjectivePoint;
function createCipherBalance(y, amount, random) {
    if (amount === 0n) {
        const L = y.multiplyUnsafe(random);
        const R = constants_1.GENERATOR.multiplyUnsafe(random);
        return { L, R };
    }
    const L = constants_1.GENERATOR.multiply(amount).add(y.multiplyUnsafe(random));
    const R = constants_1.GENERATOR.multiplyUnsafe(random);
    return { L, R };
}
/**
 * This function coincides with cairo compute_prefix
 * @param seq - Array of bigint values to hash
 * @returns The computed prefix hash
 */
function compute_prefix(seq) {
    return (0, starknet_1.poseidonHashMany)(seq);
}
/**
 * Converts a StarkPoint to a ProjectivePoint. This operation could throw an error
 * if the x and y are not coordinates from a point in the starknet curve.
 * @param {PubKey} param0 - The public key with x and y coordinates
 * @returns {ProjectivePoint} The resulting ProjectivePoint
 * @throws {Error} If the coordinates are not valid points on the starknet curve
 */
function starkPointToProjectivePoint({ x, y }) {
    return new exports.ProjectivePoint(BigInt(x), BigInt(y), 1n);
}
/**
 * Converts the ProjectivePoint to a StarkPoint
 * @param {ProjectivePoint} p - The ProjectivePoint to convert
 * @returns {StarkPoint} The resulting StarkPoint
 */
function projectivePointToStarkPoint(p) {
    const pAffine = p.toAffine();
    return { x: pAffine.x, y: pAffine.y };
}
/**
 * Constructs a public key from a given private key.
 * @param {bigint} privateKey - The private key to derive from
 * @returns {PubKey} The derived public key
 */
function derivePublicKey(privateKey) {
    return projectivePointToStarkPoint(constants_1.GENERATOR.multiply(privateKey));
}
/**
 * Converts a public key to hex format (assumes compressed format)
 * @param {PubKey} pub - The public key to convert
 * @returns {string} The hex representation of the public key
 */
function pubKeyAffineToHex(pub) {
    const point = starkPointToProjectivePoint(pub);
    return (0, utils_1.bytesToHex)(point.toRawBytes(true));
}
/**
 * Converts a public key to base58 format (assumes compressed format)
 * @param {PubKey} pub - The public key to convert
 * @returns {TongoAddress} The base58 representation as a TongoAddress
 */
function pubKeyAffineToBase58(pub) {
    const point = starkPointToProjectivePoint(pub);
    return base_1.base58.encode(point.toRawBytes(true));
}
/**
 * Converts a base58 string to affine coordinates (assumes compressed format)
 * @param {string} b58string - The base58 encoded string
 * @returns {{x: bigint, y: bigint}} The affine coordinates
 */
function pubKeyBase58ToAffine(b58string) {
    const bytes = base_1.base58.decode(b58string);
    return exports.ProjectivePoint.fromHex((0, utils_1.bytesToHex)(bytes));
}
/**
 * Converts a base58 string to hex format (assumes compressed format)
 * @param {string} b58string - The base58 encoded string
 * @returns {string} The hex representation
 */
function pubKeyBase58ToHex(b58string) {
    const bytes = base_1.base58.decode(b58string);
    return (0, utils_1.bytesToHex)(bytes);
}
/**
 * Converts a pair of StarkPoints to a CipherBalance.
 * @param {object} param0 - Object containing L and R StarkPoints
 * @param {StarkPoint} param0.L - The left point
 * @param {StarkPoint} param0.R - The right point
 * @returns {CipherBalance} The resulting CipherBalance
 */
function parseCipherBalance({ L, R }) {
    return {
        L: starkPointToProjectivePoint(L),
        R: starkPointToProjectivePoint(R),
    };
}
