// @fatsolutions/tongo-sdk/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bytesOrNumToBigInt = bytesOrNumToBigInt;
exports.isUint256 = isUint256;
exports.castBigInt = castBigInt;
exports.createCipherBalance = createCipherBalance;
exports.decipherBalance = decipherBalance;
exports.assertBalance = assertBalance;
exports.pubKeyFromSecret = pubKeyFromSecret;
const utils_1 = require("@noble/hashes/utils");
const starknet_1 = require("starknet");
const constants_1 = require("./constants");
const types_1 = require("./types");
function bytesOrNumToBigInt(x) {
    if (x instanceof Uint8Array) {
        return starknet_1.num.toBigInt("0x" + (0, utils_1.bytesToHex)(x));
    }
    else {
        return starknet_1.num.toBigInt(x);
    }
}
function isUint256(x) {
    const low = x.low;
    const high = x.high;
    return low !== undefined && high !== undefined;
}
function castBigInt(x) {
    if (starknet_1.num.isBigNumberish(x)) {
        return starknet_1.num.toBigInt(x);
    }
    else {
        return starknet_1.uint256.uint256ToBN(x);
    }
}
function createCipherBalance(y, amount, random) {
    if (amount === 0n) {
        const L = y.multiplyUnsafe(random);
        const R = constants_1.GENERATOR.multiplyUnsafe(random);
        return { L, R };
    }
    const L = constants_1.GENERATOR.multiply(amount).add(y.multiplyUnsafe(random));
    const R = constants_1.GENERATOR.multiplyUnsafe(random);
    return { L, R };
}
/**
 * Decipher the given cipher balance with the given secret key.
 * This function has to bruteforce for `b` in `g^b`. It starts at `b = 0` and
 * ends in `b = 2^32`.
 * @param {bigint} x - The secret key
 * @param {ProjectivePoint} L - The left point of the cipher balance
 * @param {ProjectivePoint} R - The right point of the cipher balance
 * @returns {bigint} The deciphered balance
 * @throws {Error} If decryption fails
 * @todo Parametrize bit size
 */
function decipherBalance(x, L, R) {
    const Rx = R.multiply(x);
    if (Rx.equals(L)) {
        return 0n;
    }
    const g_b = L.subtract(Rx);
    let b = 1n;
    let temp = constants_1.GENERATOR;
    if (temp.equals(g_b)) {
        return 1n;
    }
    while (b < 2 ** 32) {
        b = b + 1n;
        temp = temp.add(constants_1.GENERATOR);
        if (temp.equals(g_b)) {
            return b;
        }
    }
    throw new Error('Decription of Cipherbalance has failed');
}
/**
 * Asserts that the given CipherBalance is a correct encryption for the public
 * key of the given private key x and the given balance.
 * @param {bigint} x - The private key
 * @param {bigint} balance - The expected balance
 * @param {ProjectivePoint} L - The left point of the cipher balance
 * @param {ProjectivePoint} R - The right point of the cipher balance
 * @returns {boolean} True if the assertion passes, false otherwise
 */
function assertBalance(x, balance, L, R) {
    const Rx = R.multiply(x);
    const g_b = L.subtract(Rx);
    return g_b.equals(constants_1.GENERATOR.multiplyUnsafe(balance));
}
function pubKeyFromSecret(secret) {
    return (0, types_1.projectivePointToStarkPoint)(constants_1.GENERATOR.multiply(secret));
}
