// starkzap/dist/src/bridge/ethereum/oft/LayerZeroApi.js
import { getAddress, Interface } from "ethers";
import { ExternalChain } from "../../../types/index.js";
import { fromEthereumAddress } from "../../../connect/ethersRuntime.js";
const LAYERZERO_API_BASE = "https://transfer.layerzero-api.com/v1";
export class LayerZeroApi {
    constructor(config, fetcher) {
        this.config = config;
        if (fetcher) {
            this.fetcher = fetcher;
        }
        else if (typeof globalThis.fetch === "function") {
            this.fetcher = globalThis.fetch.bind(globalThis);
        }
        else {
            throw new Error("No fetch implementation available. Provide fetcher in LayerZeroApi.");
        }
    }
    async getDepositQuotes(params) {
        return this.getQuotes({
            srcChainKey: this.config.externalChainKey,
            srcTokenAddress: this.config.externalTokenAddress,
            dstChainKey: "starknet",
            dstTokenAddress: this.config.starknetTokenAddress,
            ...params,
        });
    }
    getApprovalTransaction(quotes) {
        const quote = quotes[0];
        if (!quote)
            return null;
        try {
            const approveStep = quote.userSteps.find((step) => step.description === "approve");
            return approveStep?.transaction.encoded ?? null;
        }
        catch {
            return null;
        }
    }
    getDepositTransaction(quotes) {
        return this.getBridgeTransaction(quotes);
    }
    /**
     * Extract the allowance spender from an approval transaction's calldata.
     * Parses `approve(address,uint256)` to retrieve the spender argument.
     */
    extractSpenderFromApprovalTx(approvalTx) {
        if (!approvalTx?.data)
            return null;
        try {
            const approveInterface = new Interface([
                "function approve(address spender, uint256 value)",
            ]);
            const decoded = approveInterface.parseTransaction({
                data: approvalTx.data,
            });
            const spender = decoded?.args[0];
            return spender ? fromEthereumAddress(spender, { getAddress }) : null;
        }
        catch {
            return null;
        }
    }
    async getQuotes(params) {
        const response = await this.fetcher(`${LAYERZERO_API_BASE}/quotes`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                ...(this.config.apiKey ? { "x-api-key": this.config.apiKey } : {}),
            },
            body: JSON.stringify({
                ...params,
                amount: params.amount.toBase().toString(),
                options: {
                    amountType: "EXACT_SRC_AMOUNT",
                    feeTolerance: { type: "PERCENT", amount: 1 },
                    dstNativeDropAmount: "0",
                },
            }),
        });
        if (!response.ok) {
            throw new Error(`LayerZero API request failed (${response.status}): ${await response.text()}`);
        }
        const data = (await response.json());
        return data.quotes;
    }
    getBridgeTransaction(quotes) {
        const quote = quotes[0];
        if (!quote)
            return null;
        try {
            const bridgeStep = quote.userSteps.find((step) => step.description === "bridge");
            return bridgeStep?.transaction.encoded ?? null;
        }
        catch {
            return null;
        }
    }
}
