// starkzap/dist/src/bridge/ethereum/oft/OftBridge.js
import { EthereumBridge } from "../../../bridge/ethereum/EthereumBridge.js";
import { Amount, EthereumBridgeToken, ExternalChain, fromAddress, } from "../../../types/index.js";
import { FeeErrorCause } from "../../../types/errors.js";
import { LayerZeroApi } from "../../../bridge/ethereum/oft/LayerZeroApi.js";
import { DEFAULT_OFT_DEPOSIT_GAS_REQUIREMENT, DEFAULT_OFT_MIN_AMOUNT, OFT_MIN_AMOUNT_BY_TOKEN_ID, } from "../../../bridge/ethereum/oft/constants.js";
const DUMMY_SN_ADDRESS = fromAddress("0x023123100123103023123acb1231231231231031231ca123f23123123123100a");
const DUMMY_ETH_ADDRESS = "0x0000000000000000000000000000000000000001";
const DUMMY_DEPOSIT_TX_CACHE_TTL_MS = 60000;
export class OftBridge extends EthereumBridge {
    constructor(bridgeToken, config, starknetWallet, apiKey) {
        super(bridgeToken, config, starknetWallet, []);
        this.dummyDepositTxCache = null;
        const chainId = starknetWallet.getChainId();
        if (!chainId.isMainnet()) {
            throw new Error("OFT bridging is only supported on Starknet Mainnet. " +
                "The LayerZero Value Transfer API does not support testnets.");
        }
        this.layerZeroApi = new LayerZeroApi({
            externalTokenAddress: bridgeToken.address,
            starknetTokenAddress: bridgeToken.starknetAddress,
            externalChainKey: ExternalChain.ETHEREUM,
            apiKey,
        });
    }
    async deposit(recipient, amount, _options) {
        await this.approveSpendingOf(amount);
        const signerAddress = await this.config.signer.getAddress();
        const quotes = await this.layerZeroApi.getDepositQuotes({
            srcWalletAddress: signerAddress,
            dstWalletAddress: recipient,
            amount: amount,
        });
        const depositTx = this.layerZeroApi.getDepositTransaction(quotes);
        if (!depositTx) {
            throw new Error("Failed to get OFT deposit transaction from LayerZero API.");
        }
        const response = await this.execute(depositTx);
        this.clearCachedAllowance();
        this.clearDummyDepositTxCache();
        return { hash: response.hash };
    }
    async getDepositFeeEstimate(_options) {
        const [allowance, dummyDepositTx, approvalFeeEstimation] = await Promise.all([
            this.getAllowance(),
            this.getDummyDepositTx(),
            this.estimateApprovalFee(),
        ]);
        const { approvalFee, approvalFeeError } = approvalFeeEstimation;
        const interchainFee = this.ethAmount(BigInt(dummyDepositTx?.value?.toString() ?? "0"));
        let l1Fee;
        let l1FeeError;
        const needsFallback = allowance !== null && allowance.isZero();
        if (needsFallback || !dummyDepositTx) {
            const feeDecimal = DEFAULT_OFT_DEPOSIT_GAS_REQUIREMENT *
                (await this.getEthereumGasPrice());
            l1Fee = this.ethAmount(feeDecimal);
            if (!dummyDepositTx) {
                l1FeeError = FeeErrorCause.GENERIC_L1_FEE_ERROR;
            }
        }
        else {
            try {
                const [gasUnits, gasPrice] = await Promise.all([
                    this.config.provider.estimateGas(dummyDepositTx),
                    this.getEthereumGasPrice(),
                ]);
                l1Fee = this.ethAmount(gasUnits * gasPrice);
            }
            catch {
                l1Fee = this.ethAmount(0n);
                l1FeeError = FeeErrorCause.GENERIC_L1_FEE_ERROR;
            }
        }
        return {
            l1Fee,
            l1FeeError,
            l2Fee: interchainFee,
            interchainFee,
            approvalFee,
            approvalFeeError,
        };
    }
    async getAllowanceSpender() {
        if (this.cachedSpender !== undefined) {
            return this.cachedSpender;
        }
        try {
            const quotes = await this.layerZeroApi.getDepositQuotes({
                srcWalletAddress: DUMMY_ETH_ADDRESS,
                dstWalletAddress: DUMMY_SN_ADDRESS.toString(),
                amount: this.getOftMinAmount(),
            });
            const approvalTx = this.layerZeroApi.getApprovalTransaction(quotes);
            this.cachedSpender =
                this.layerZeroApi.extractSpenderFromApprovalTx(approvalTx);
        }
        catch {
            // Do not cache transient API failures as "no spender";
            // allow subsequent calls to retry spender discovery.
            this.cachedSpender = undefined;
            return null;
        }
        return this.cachedSpender;
    }
    async getEthereumGasPrice() {
        const gasData = await this.config.provider.getFeeData();
        const gasPrice = gasData.gasPrice ?? 0n;
        const maxFeePerGas = gasData.maxFeePerGas;
        return maxFeePerGas && gasData.maxPriorityFeePerGas
            ? maxFeePerGas
            : gasPrice;
    }
    getOftMinAmount() {
        const amountBase = OFT_MIN_AMOUNT_BY_TOKEN_ID[this.bridgeToken.id] ?? DEFAULT_OFT_MIN_AMOUNT;
        return Amount.fromRaw(amountBase, this.bridgeToken.decimals, this.bridgeToken.symbol);
    }
    getDummyDepositTx() {
        const now = Date.now();
        const cached = this.dummyDepositTxCache;
        if (cached && now - cached.createdAt < DUMMY_DEPOSIT_TX_CACHE_TTL_MS) {
            return cached.promise;
        }
        const promise = this.fetchDummyDepositTx().then((tx) => {
            // Retry quickly after transient failures rather than pinning null for the full TTL.
            if (!tx && this.dummyDepositTxCache?.promise === promise) {
                this.dummyDepositTxCache = null;
            }
            return tx;
        });
        this.dummyDepositTxCache = { promise, createdAt: now };
        return promise;
    }
    async fetchDummyDepositTx() {
        try {
            const signerAddress = await this.config.signer.getAddress();
            const quotes = await this.layerZeroApi.getDepositQuotes({
                srcWalletAddress: signerAddress,
                dstWalletAddress: DUMMY_SN_ADDRESS.toString(),
                amount: this.getOftMinAmount(),
            });
            return this.layerZeroApi.getDepositTransaction(quotes);
        }
        catch {
            return null;
        }
    }
    clearDummyDepositTxCache() {
        this.dummyDepositTxCache = null;
    }
}
