// @fatsolutions/tongo-sdk/dist/operations/fund.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FundOperation = void 0;
const starknet_1 = require("starknet");
const utils_js_1 = require("../utils.js");
const operation_js_1 = require("./operation.js");
class FundOperation {
    type;
    Tongo;
    to;
    amount;
    hint;
    proof;
    auditPart;
    approve;
    constructor({ to, amount, proof, auditPart, Tongo, hint }) {
        this.type = operation_js_1.OperationType.Fund;
        this.to = to;
        this.amount = amount;
        this.hint = hint;
        this.auditPart = auditPart;
        this.proof = proof;
        this.Tongo = Tongo;
    }
    toCalldata() {
        return this.Tongo.populate("fund", [
            {
                to: this.to,
                amount: this.amount,
                hint: this.hint,
                proof: this.proof,
                auditPart: this.auditPart,
            },
        ]);
    }
    // TODO: better ux for this. Maybe return the call?
    async populateApprove() {
        const erc20 = await this.Tongo.ERC20();
        const erc20_addres = starknet_1.num.toHex(erc20);
        const tongo_address = this.Tongo.address;
        const rate = await this.Tongo.get_rate();
        const amount = starknet_1.cairo.uint256(this.amount * (0, utils_js_1.castBigInt)(rate));
        const calldata = starknet_1.CallData.compile({ spender: tongo_address, amount: amount });
        this.approve = { contractAddress: erc20_addres, entrypoint: "approve", calldata };
    }
}
exports.FundOperation = FundOperation;
