// @fatsolutions/tongo-sdk/dist/operations/transfer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferOperation = void 0;
const operation_js_1 = require("./operation.js");
class TransferOperation {
    type = operation_js_1.OperationType.Transfer;
    Tongo;
    from;
    to;
    transferBalance;
    transferBalanceSelf;
    auxiliarCipher;
    auxiliarCipher2;
    hintTransfer;
    hintLeftover;
    proof;
    auditPart;
    auditPartTransfer;
    constructor({ from, to, transferBalance, transferBalanceSelf, proof, auditPart, auditPartTransfer, auxiliarCipher, auxiliarCipher2, Tongo, hintTransfer, hintLeftover, }) {
        this.from = from;
        this.to = to;
        this.transferBalance = transferBalance;
        this.transferBalanceSelf = transferBalanceSelf;
        this.auxiliarCipher = auxiliarCipher;
        this.auxiliarCipher2 = auxiliarCipher2;
        this.hintTransfer = hintTransfer;
        this.hintLeftover = hintLeftover;
        this.proof = proof;
        this.auditPart = auditPart;
        this.auditPartTransfer = auditPartTransfer;
        this.Tongo = Tongo;
    }
    toCalldata() {
        return this.Tongo.populate("transfer", [
            {
                from: this.from,
                to: this.to,
                transferBalance: this.transferBalance,
                transferBalanceSelf: this.transferBalanceSelf,
                auxiliarCipher: this.auxiliarCipher,
                auxiliarCipher2: this.auxiliarCipher2,
                hintTransfer: this.hintTransfer,
                hintLeftover: this.hintLeftover,
                proof: this.proof,
                auditPart: this.auditPart,
                auditPartTransfer: this.auditPartTransfer,
            },
        ]);
    }
}
exports.TransferOperation = TransferOperation;
