// @fatsolutions/tongo-sdk/dist/operations/withdraw.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WithdrawOperation = void 0;
const starknet_1 = require("starknet");
const operation_js_1 = require("./operation.js");
class WithdrawOperation {
    type = operation_js_1.OperationType.Withdraw;
    from;
    to;
    amount;
    hint;
    auxiliarCipher;
    proof;
    auditPart;
    Tongo;
    constructor({ from, to, amount, proof, auditPart, Tongo, hint, auxiliarCipher }) {
        this.from = from;
        this.to = to;
        this.amount = amount;
        this.auxiliarCipher = auxiliarCipher;
        this.hint = hint;
        this.proof = proof;
        this.auditPart = auditPart;
        this.Tongo = Tongo;
    }
    toCalldata() {
        return this.Tongo.populate("withdraw", [
            {
                from: this.from,
                amount: this.amount,
                hint: this.hint,
                to: starknet_1.num.toHex(this.to),
                auxiliarCipher: this.auxiliarCipher,
                auditPart: this.auditPart,
                proof: this.proof,
            },
        ]);
    }
}
exports.WithdrawOperation = WithdrawOperation;
