// starkzap/dist/src/providers/registry.js
/**
 * Generic, type-safe provider registry used by Swap, DCA, and Lending modules.
 *
 * Handles registration, default selection, lookup, and listing — the logic
 * that was previously duplicated across three separate implementations.
 */
export class ProviderRegistry {
    constructor(domain) {
        this.domain = domain;
        this.providers = new Map();
        this.defaultId = null;
    }
    /** Register a provider. Optionally make it the default. */
    register(provider, makeDefault = false) {
        this.providers.set(provider.id, provider);
        if (makeDefault || this.defaultId == null) {
            this.defaultId = provider.id;
        }
    }
    /** Set the default provider by id. Throws if the id is not registered. */
    setDefault(id) {
        this.get(id); // validates existence
        this.defaultId = id;
    }
    /** Look up a provider by id. Throws if not found. */
    get(id) {
        const provider = this.providers.get(id);
        if (!provider) {
            throw new Error(`Unknown ${this.domain} provider "${id}". Registered providers: ${this.list().join(", ")}`);
        }
        return provider;
    }
    /** Return the default provider. Throws if none is configured. */
    getDefault() {
        if (!this.defaultId) {
            throw new Error(`No default ${this.domain} provider configured`);
        }
        return this.get(this.defaultId);
    }
    /** List all registered provider ids. */
    list() {
        return Array.from(this.providers.keys());
    }
}
