// starkzap/dist/src/connect/evm.js
import { describeValue } from "../connect/utils.js";
import { ExternalChain } from "../types/index.js";
import { loadEthers } from "../connect/ethersRuntime.js";
import { fromEthereumAddress } from "../connect/ethersRuntime.js";
function assertEip1193Provider(provider) {
    if (typeof provider === "object" &&
        provider !== null &&
        "request" in provider &&
        typeof provider.request === "function") {
        return provider;
    }
    throw new Error(`EVM provider must implement EIP-1193 request(). Received ${describeValue(provider)}.`);
}
export var EthereumNetwork;
(function (EthereumNetwork) {
    EthereumNetwork[EthereumNetwork["MAINNET"] = 1] = "MAINNET";
    EthereumNetwork[EthereumNetwork["SEPOLIA"] = 11155111] = "SEPOLIA";
})(EthereumNetwork || (EthereumNetwork = {}));
export class ConnectedEthereumWallet {
    constructor(address, provider, network) {
        this.address = address;
        this.provider = provider;
        this.network = network;
        this.chain = ExternalChain.ETHEREUM;
    }
    async toEthWalletConfig(ethereumRpcUrl) {
        const ethers = await loadEthers("ConnectedEthereumWallet.toEthWalletConfig");
        const ethChainIdRaw = await this.provider.request({
            method: "eth_chainId",
        });
        const ethChainId = Number(BigInt(ethChainIdRaw));
        const networkId = this.network;
        if (ethChainId !== networkId) {
            throw new Error(`Cannot create Ethereum Bridge. Expected ethereum chain id to be ${networkId} but got ${ethChainId}.`);
        }
        const browserProvider = new ethers.BrowserProvider(this.provider, networkId);
        const signer = await browserProvider.getSigner(this.address);
        let provider;
        if (ethereumRpcUrl) {
            const rpcProvider = new ethers.JsonRpcProvider(ethereumRpcUrl, networkId);
            const rpcNetwork = await rpcProvider.getNetwork();
            const rpcChainId = Number(rpcNetwork.chainId);
            if (rpcChainId !== networkId) {
                throw new Error(`Custom Ethereum RPC URL is on chain ${rpcChainId} but the connected wallet is on chain ${networkId}.`);
            }
            provider = rpcProvider;
        }
        else {
            provider = browserProvider;
        }
        return { provider, signer };
    }
    static async from(options, starknetChain) {
        const ethers = await loadEthers("ConnectedEthereumWallet.from");
        const address = fromEthereumAddress(options.address, ethers);
        const provider = assertEip1193Provider(options.provider);
        const numericChainId = typeof options.chainId === "string"
            ? Number(options.chainId)
            : options.chainId;
        if (!Number.isFinite(numericChainId) || numericChainId <= 0) {
            throw new Error(`Invalid EVM chain ID: ${String(options.chainId)}`);
        }
        let network;
        if (numericChainId === EthereumNetwork.MAINNET) {
            network = EthereumNetwork.MAINNET;
        }
        else if (numericChainId === EthereumNetwork.SEPOLIA) {
            network = EthereumNetwork.SEPOLIA;
        }
        else {
            throw new Error(`Unsupported chainId ${numericChainId} for Ethereum`);
        }
        if (network === EthereumNetwork.MAINNET && !starknetChain.isMainnet()) {
            throw new Error(`Ethereum Mainnet cannot be used with Starknet Sepolia.`);
        }
        if (network === EthereumNetwork.SEPOLIA && !starknetChain.isSepolia()) {
            throw new Error("Ethereum Sepolia cannot be used with Starknet Mainnet.");
        }
        return new ConnectedEthereumWallet(address, provider, network);
    }
}
