// starkzap/dist/src/connect/solana.js
import { assertNonEmptyString, describeValue } from "../connect/utils.js";
import { ExternalChain } from "../types/index.js";
import { loadSolanaWeb3 } from "../connect/solanaWeb3Runtime.js";
import { fromSolanaAddress } from "../types/solanaAddress.js";
export var SolanaNetwork;
(function (SolanaNetwork) {
    SolanaNetwork["MAINNET"] = "5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp";
    SolanaNetwork["TESTNET"] = "4uhcVJyU9pJkvQyS88uRDiswHXSCkY3z";
})(SolanaNetwork || (SolanaNetwork = {}));
function assertSolanaProvider(signer) {
    if (typeof signer === "object" &&
        signer !== null &&
        "signAndSendTransaction" in signer &&
        typeof signer.signAndSendTransaction === "function") {
        return signer;
    }
    throw new Error(`Solana signer must implement signAndSendTransaction(). Received ${describeValue(signer)}.`);
}
export class ConnectedSolanaWallet {
    constructor(address, provider, network) {
        this.address = address;
        this.provider = provider;
        this.network = network;
        this.chain = ExternalChain.SOLANA;
    }
    static async from(options, starknetChain) {
        const solanaWeb3 = await loadSolanaWeb3("ConnectedSolanaWallet.from");
        const chainId = assertNonEmptyString(options.chainId, "chainId");
        const signer = assertSolanaProvider(options.provider);
        const address = fromSolanaAddress(options.address, solanaWeb3);
        let network;
        if (chainId === SolanaNetwork.MAINNET) {
            network = SolanaNetwork.MAINNET;
        }
        else if (chainId === SolanaNetwork.TESTNET) {
            network = SolanaNetwork.TESTNET;
        }
        else {
            throw new Error(`Unsupported chainId ${chainId} for Solana`);
        }
        if (network === SolanaNetwork.MAINNET && !starknetChain.isMainnet()) {
            throw new Error("Solana Mainnet cannot be used with Starknet Sepolia.");
        }
        if (network === SolanaNetwork.TESTNET && !starknetChain.isSepolia()) {
            throw new Error("Solana Testnet cannot be used with Starknet Mainnet.");
        }
        return new ConnectedSolanaWallet(address, signer, network);
    }
}
