// starkzap/dist/src/lending/client.js
import { VesuLendingProvider } from "../lending/vesu/index.js";
import { assertLendingContext, hydrateBorrowRequest, hydrateDepositRequest, hydrateHealthRequest, hydrateRepayRequest, hydrateWithdrawMaxRequest, hydrateWithdrawRequest, resolveLendingSource, stripProvider, } from "../lending/utils.js";
export class LendingClient {
    constructor(context, defaultProvider) {
        this.defaultProviderId = null;
        this.context = context;
        this.providers = new Map();
        this.registerProvider(defaultProvider ?? new VesuLendingProvider(), true);
    }
    registerProvider(provider, makeDefault = false) {
        this.providers.set(provider.id, provider);
        if (makeDefault || this.defaultProviderId == null) {
            this.defaultProviderId = provider.id;
        }
    }
    setDefaultProvider(providerId) {
        if (!this.providers.has(providerId)) {
            throw new Error(`Unknown lending provider "${providerId}". Registered providers: ${this.listProviders().join(", ")}`);
        }
        this.defaultProviderId = providerId;
    }
    getLendingProvider(providerId) {
        const provider = this.providers.get(providerId);
        if (!provider) {
            throw new Error(`Unknown lending provider "${providerId}". Registered providers: ${this.listProviders().join(", ")}`);
        }
        return provider;
    }
    getDefaultLendingProvider() {
        if (!this.defaultProviderId) {
            throw new Error("No default lending provider configured");
        }
        return this.getLendingProvider(this.defaultProviderId);
    }
    listProviders() {
        return Array.from(this.providers.keys());
    }
    async getMarkets(request = {}) {
        return await this.resolveRequestProvider(request.provider).getMarkets(this.context.getChainId());
    }
    async getPositions(request = {}) {
        const provider = this.resolveRequestProvider(request.provider);
        if (!provider.getPositions) {
            throw new Error(`Lending provider "${provider.id}" does not support position queries`);
        }
        return await provider.getPositions(this.providerContext(), {
            user: request.user ?? this.context.address,
        });
    }
    async getMaxBorrowAmount(request) {
        const provider = this.resolveRequestProvider(request.provider);
        if (!provider.getMaxBorrowAmount) {
            throw new Error(`Lending provider "${provider.id}" does not support max borrow queries`);
        }
        return await provider.getMaxBorrowAmount(this.providerContext(), {
            ...stripProvider(request),
            user: request.user ?? this.context.address,
        });
    }
    async prepareDeposit(request) {
        return await this.prepareWithProvider(request, hydrateDepositRequest, (provider, context, hydrated) => provider.prepareDeposit(context, hydrated));
    }
    async deposit(request, options) {
        return await this.executePrepared(this.prepareDeposit(request), options);
    }
    async prepareWithdraw(request) {
        return await this.prepareWithProvider(request, hydrateWithdrawRequest, (provider, context, hydrated) => provider.prepareWithdraw(context, hydrated));
    }
    async withdraw(request, options) {
        return await this.executePrepared(this.prepareWithdraw(request), options);
    }
    async prepareWithdrawMax(request) {
        return await this.prepareWithProvider(request, hydrateWithdrawMaxRequest, (provider, context, hydrated) => {
            if (!provider.prepareWithdrawMax) {
                throw new Error(`Lending provider "${provider.id}" does not support max-withdraw`);
            }
            return provider.prepareWithdrawMax(context, hydrated);
        });
    }
    async withdrawMax(request, options) {
        return await this.executePrepared(this.prepareWithdrawMax(request), options);
    }
    async prepareBorrow(request) {
        return await this.prepareWithProvider(request, hydrateBorrowRequest, (provider, context, hydrated) => provider.prepareBorrow(context, hydrated));
    }
    async borrow(request, options) {
        return await this.executePrepared(this.prepareBorrow(request), options);
    }
    async prepareRepay(request) {
        return await this.prepareWithProvider(request, hydrateRepayRequest, (provider, context, hydrated) => provider.prepareRepay(context, hydrated));
    }
    async repay(request, options) {
        return await this.executePrepared(this.prepareRepay(request), options);
    }
    async getPosition(request) {
        const provider = this.resolveRequestProvider(request.provider);
        return await provider.getPosition(this.providerContext(), {
            ...stripProvider(request),
            user: request.user ?? this.context.address,
        });
    }
    async getHealth(request) {
        const provider = this.resolveRequestProvider(request.provider);
        return await provider.getHealth(this.providerContext(), {
            ...hydrateHealthRequest(request, this.context.address),
        });
    }
    async quoteHealth(request) {
        this.assertQuoteHealthCompatibility(request);
        const current = await this.getHealth(request.health);
        const prepared = await this.prepareAction(request.action);
        const simulation = await this.context.preflight({
            calls: prepared.calls,
            ...(request.feeMode != null && { feeMode: request.feeMode }),
        });
        const projected = await this.projectHealth(request, current);
        return { current, prepared, simulation, projected };
    }
    async prepareAction(input) {
        if (input.action === "deposit") {
            return await this.prepareDeposit(input.request);
        }
        if (input.action === "withdraw") {
            return await this.prepareWithdraw(input.request);
        }
        if (input.action === "borrow") {
            return await this.prepareBorrow(input.request);
        }
        return await this.prepareRepay(input.request);
    }
    async projectHealth(request, current) {
        const [healthProvider, actionProvider] = this.resolveQuoteProviders(request);
        if (healthProvider !== actionProvider) {
            return null;
        }
        if (!actionProvider.quoteProjectedHealth) {
            return null;
        }
        return await actionProvider.quoteProjectedHealth(this.providerContext(), request, current);
    }
    assertQuoteHealthCompatibility(request) {
        const [healthProvider, actionProvider] = this.resolveQuoteProviders(request);
        if (healthProvider !== actionProvider) {
            throw new Error("quoteHealth requires action and health to use the same lending provider");
        }
        if (request.action.action !== "borrow" &&
            request.action.action !== "repay") {
            return;
        }
        const healthRequest = hydrateHealthRequest(request.health, this.context.address);
        const actionRequest = request.action.action === "borrow"
            ? hydrateBorrowRequest(request.action.request, this.context.address)
            : hydrateRepayRequest(request.action.request, this.context.address);
        if (actionRequest.poolAddress !== healthRequest.poolAddress ||
            actionRequest.user !== healthRequest.user ||
            actionRequest.collateralToken.address !==
                healthRequest.collateralToken.address ||
            actionRequest.debtToken.address !== healthRequest.debtToken.address) {
            throw new Error("quoteHealth requires action and health to target the same lending position");
        }
    }
    resolveRequestProvider(source) {
        const provider = resolveLendingSource(source, this);
        assertLendingContext(provider, this.context.getChainId());
        return provider;
    }
    resolveQuoteProviders(request) {
        return [
            this.resolveRequestProvider(request.health.provider),
            this.resolveRequestProvider(request.action.request.provider),
        ];
    }
    async prepareWithProvider(request, hydrate, prepare) {
        const provider = this.resolveRequestProvider(request.provider);
        const prepared = await prepare(provider, this.providerContext(), hydrate(request, this.context.address));
        this.assertPreparedCalls(prepared, provider.id);
        return prepared;
    }
    async executePrepared(preparedPromise, options) {
        return await this.context.execute((await preparedPromise).calls, options);
    }
    providerContext() {
        return {
            chainId: this.context.getChainId(),
            provider: this.context.getProvider(),
            walletAddress: this.context.address,
        };
    }
    assertPreparedCalls(prepared, providerId) {
        if (prepared.calls.length > 0) {
            return;
        }
        throw new Error(`Lending provider "${providerId}" returned no calls`);
    }
}
