// starkzap/dist/src/lending/utils.js
export function resolveLendingSource(source, resolver) {
    if (source == null) {
        return resolver.getDefaultLendingProvider();
    }
    if (typeof source === "string") {
        return resolver.getLendingProvider(source);
    }
    return source;
}
export function assertLendingContext(provider, walletChainId) {
    const chain = walletChainId.toLiteral();
    if (!provider.supportsChain(walletChainId)) {
        throw new Error(`Lending provider "${provider.id}" does not support chain "${chain}"`);
    }
}
export function stripProvider(request) {
    const { provider: _provider, ...rest } = request;
    return rest;
}
export function hydrateDepositRequest(request, walletAddress) {
    return {
        ...stripProvider(request),
        receiver: request.receiver ?? walletAddress,
    };
}
export function hydrateWithdrawRequest(request, walletAddress) {
    return {
        ...stripProvider(request),
        receiver: request.receiver ?? walletAddress,
        owner: request.owner ?? walletAddress,
    };
}
export function hydrateWithdrawMaxRequest(request, walletAddress) {
    return {
        ...stripProvider(request),
        receiver: request.receiver ?? walletAddress,
        owner: request.owner ?? walletAddress,
    };
}
export function hydrateBorrowRequest(request, walletAddress) {
    return {
        ...stripProvider(request),
        user: request.user ?? walletAddress,
        collateralDenomination: request.collateralDenomination ?? "assets",
        debtDenomination: request.debtDenomination ?? "assets",
    };
}
export function hydrateRepayRequest(request, walletAddress) {
    return {
        ...stripProvider(request),
        user: request.user ?? walletAddress,
        collateralDenomination: request.collateralDenomination ?? "assets",
        withdrawCollateral: request.withdrawCollateral ?? false,
        debtDenomination: request.debtDenomination ?? "assets",
    };
}
export function hydrateHealthRequest(request, walletAddress) {
    return {
        ...stripProvider(request),
        user: request.user ?? walletAddress,
    };
}
