// @fatsolutions/tongo-sdk/dist/provers/ragequit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RAGEQUIT_CAIRO_STRING = void 0;
exports.proveRagequit = proveRagequit;
exports.verifyRagequit = verifyRagequit;
const protocols_1 = require("@fatsolutions/she/protocols");
const constants_1 = require("../constants");
const types_1 = require("../types");
const utils_1 = require("../utils");
const she_1 = require("@fatsolutions/she");
// cairo string 'ragequit'
exports.RAGEQUIT_CAIRO_STRING = 8241982478457596276n;
/**
 * Computes the prefix by hashing some public inputs.
 * @param {InputsRagequit} inputs - The ragequit operation inputs
 * @returns {bigint} The computed prefix hash
 */
function prefixRagequit(inputs) {
    const { chain_id, tongo_address, sender_address } = inputs.prefix_data;
    const seq = [
        chain_id,
        tongo_address,
        sender_address,
        exports.RAGEQUIT_CAIRO_STRING,
        inputs.y.toAffine().x,
        inputs.y.toAffine().y,
        inputs.nonce,
        inputs.amount,
        inputs.to,
        inputs.currentBalance.L.toAffine().x,
        inputs.currentBalance.L.toAffine().y,
        inputs.currentBalance.R.toAffine().x,
        inputs.currentBalance.R.toAffine().y,
    ];
    return (0, types_1.compute_prefix)(seq);
}
function proveRagequit(private_key, initial_cipherbalance, nonce, to, full_amount, prefix_data) {
    const x = private_key;
    const y = constants_1.GENERATOR.multiply(x);
    const { L: L0, R: R0 } = initial_cipherbalance;
    //this is to assert that storedbalance is an encription of the balance amount
    const g_b = L0.subtract(R0.multiplyUnsafe(x));
    const temp = constants_1.GENERATOR.multiplyUnsafe(full_amount);
    if (!g_b.equals(temp)) {
        throw new Error("storedBalance is not an encryption of balance");
    }
    ;
    const inputs = {
        y,
        nonce,
        to,
        amount: full_amount,
        currentBalance: initial_cipherbalance,
        prefix_data,
    };
    const prefix = prefixRagequit(inputs);
    const kx = (0, she_1.generateRandom)();
    const Ax = constants_1.GENERATOR.multiply(kx);
    const AR = R0.multiplyUnsafe(kx);
    const c = (0, she_1.compute_challenge)(prefix, [Ax, AR]);
    const sx = (0, she_1.compute_s)(kx, x, c);
    const proof = { Ax, AR, sx };
    const newBalance = (0, utils_1.createCipherBalance)(y, 0n, 1n);
    return { inputs, proof, newBalance };
}
/**
 * Verifies the ragequit operation. First, users have to show knowledge of the private key. Then, users have to provide
 * a cleartext of the amount b stored in their balances. The contract will construct a cipher balance
 * (L2, R2) = (g**b y, g) with randomness r=1. Users have to provide a zk proof that (L2,R2) is encrypting
 * the same amount that the stored cipher balance (L1,R1). This is done by noting that
 * L1/L2 = y**r1/y**r2 = (R1/R2)**x. We need to prove a poe for Y=G**x with Y=L1/L2 and G=R1/R2
 *
 * Protocol:
 * - P:  k <-- R        sends    A=G**k
 * - V:  c <-- R        sends    c
 * - P:  s = k + c*x    send     s
 *
 * The verifier asserts:
 * - G**sr == A * (Y**c)
 *
 * Complexity:
 * - EC_MUL: 7
 * - EC_ADD: 5
 *
 * @param {InputsRagequit} inputs - The ragequit operation inputs
 * @param {ProofOfRagequit} proof - The proof to verify
 * @returns {boolean} True if the proof is valid, false otherwise
 */
function verifyRagequit(inputs, proof) {
    const prefix = prefixRagequit(inputs);
    const c = (0, she_1.compute_challenge)(prefix, [proof.Ax, proof.AR]);
    let res = protocols_1.poe._verify(inputs.y, constants_1.GENERATOR, proof.Ax, c, proof.sx);
    if (res == false) {
        throw new Error("error in poe y");
    }
    const { L: L1, R: R1 } = inputs.currentBalance;
    const L = L1.subtract(constants_1.GENERATOR.multiply(inputs.amount));
    res = protocols_1.poe._verify(L, R1, proof.AR, c, proof.sx);
    if (res == false) {
        throw new Error("error in poe R");
    }
}
