// @fatsolutions/tongo-sdk/dist/provers/withdraw.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WITHDRAW_CAIRO_STRING = void 0;
exports.proveWithdraw = proveWithdraw;
exports.verifyWithdraw = verifyWithdraw;
const she_1 = require("@fatsolutions/she");
const protocols_1 = require("@fatsolutions/she/protocols");
const protocols_2 = require("@fatsolutions/she/protocols");
const constants_1 = require("../constants");
const range_1 = require("../provers/range");
const types_1 = require("../types");
const utils_1 = require("../utils");
// cairo string 'withdraw'
exports.WITHDRAW_CAIRO_STRING = 8604536554778681719n;
/**
 * Computes the prefix by hashing some public inputs.
 * @param {InputsWithdraw} inputs - The inputs from the proof
 * @returns {bigint} The computed prefix hash
 */
function prefixWithdraw(inputs) {
    const { chain_id, tongo_address, sender_address } = inputs.prefix_data;
    const { L: L0, R: R0 } = inputs.currentBalance;
    const { L: V, R: R_aux } = inputs.auxiliarCipher;
    const seq = [
        chain_id,
        tongo_address,
        sender_address,
        exports.WITHDRAW_CAIRO_STRING,
        inputs.y.toAffine().x,
        inputs.y.toAffine().y,
        inputs.nonce,
        inputs.amount,
        inputs.to,
        L0.toAffine().x,
        L0.toAffine().y,
        R0.toAffine().x,
        R0.toAffine().y,
        V.toAffine().x,
        V.toAffine().y,
        R_aux.toAffine().x,
        R_aux.toAffine().y,
    ];
    return (0, types_1.compute_prefix)(seq);
}
function proveWithdraw(private_key, initial_balance, amount, to, initial_cipherbalance, nonce, bit_size, prefix_data) {
    const x = private_key;
    const y = constants_1.GENERATOR.multiply(x);
    const { L: L0, R: R0 } = initial_cipherbalance;
    //this is to assert that storedbalance is an encryption of the balance amount
    const g_b = L0.subtract(R0.multiplyUnsafe(x));
    const temp = constants_1.GENERATOR.multiplyUnsafe(initial_balance);
    if (!g_b.equals(temp)) {
        throw new Error("storedBalance is not an encryption of balance");
    }
    ;
    const left = initial_balance - amount;
    // This precomputation is usefull to know add R_aux and V to the prefix computation
    const { randomness, total_random } = protocols_2.range.pregenerate_randomness(bit_size);
    const auxiliarCipher = (0, utils_1.createCipherBalance)(constants_1.SECONDARY_GENERATOR, left, total_random);
    const inputs = {
        y,
        nonce,
        currentBalance: initial_cipherbalance,
        to,
        amount,
        bit_size,
        auxiliarCipher,
        prefix_data,
    };
    const prefix = prefixWithdraw(inputs);
    const { r, range } = (0, range_1.generateRangeProof)(left, bit_size, randomness, prefix);
    if (r !== total_random) {
        throw new Error("random mismatch");
    }
    ;
    const kb = (0, she_1.generateRandom)();
    const kx = (0, she_1.generateRandom)();
    const kr = (0, she_1.generateRandom)();
    const A_x = constants_1.GENERATOR.multiplyUnsafe(kx);
    const A_r = constants_1.GENERATOR.multiplyUnsafe(kr);
    const A = constants_1.GENERATOR.multiplyUnsafe(kb).add(R0.multiplyUnsafe(kx));
    const A_v = constants_1.GENERATOR.multiplyUnsafe(kb).add(constants_1.SECONDARY_GENERATOR.multiplyUnsafe(kr));
    const c = (0, she_1.compute_challenge)(prefix, [A_x, A_r, A, A_v]);
    const sb = (0, she_1.compute_s)(kb, left, c);
    const sx = (0, she_1.compute_s)(kx, x, c);
    const sr = (0, she_1.compute_s)(kr, r, c);
    const proof = {
        A_x,
        A_r,
        A,
        A_v,
        sx,
        sb,
        sr,
        range,
    };
    // compute the cipherbalance that y will have at the end of the withdraw
    const cipher = (0, utils_1.createCipherBalance)(y, amount, exports.WITHDRAW_CAIRO_STRING);
    const newBalance = { L: L0.subtract(cipher.L), R: R0.subtract(cipher.R) };
    return { inputs, proof, newBalance };
}
/**
 * Verifies the withdraw operation. First, users have to show knowledge of the private key. Then, users have to provide
 * a cleartext of the amount b to withdraw. The contract will construct a cipher balance (L2, R2) = (g**b y**r2, g**r2)
 * with randomness r2='withdraw'. The contract will subtract (L2,R2) to the stored balance of the user. The user have
 * to provide a zk proof that the final cipher balance is encrypting a positive (a value in (0, u**32)) amount b_left. To do
 * this when the RangeProof is verified, it returns a V = g**b_left h**r, with b_left positive. V is used as an L part of
 * a cipher balance, users have to prove that the cipher balance (V, R_aux = g**r) is encrypting the same amount
 * that the final cipher balance.
 *
 * Complexity:
 * - EC_MUL: 12 + n*5 = 172 for u32
 * - EC_ADD: 8 + n*4  = 136 for u32
 *
 * @param {InputsWithdraw} inputs - The withdraw operation inputs
 * @param {ProofOfWithdraw} proof - The proof to verify
 * @returns {boolean} True if the proof is valid, false otherwise
 */
function verifyWithdraw(inputs, proof) {
    const bit_size = inputs.bit_size;
    const prefix = prefixWithdraw(inputs);
    const c = (0, she_1.compute_challenge)(prefix, [proof.A_x, proof.A_r, proof.A, proof.A_v]);
    let { L: L0, R: R0 } = inputs.currentBalance;
    let { L: V, R: R_aux } = inputs.auxiliarCipher;
    L0 = L0.subtract(constants_1.GENERATOR.multiply(inputs.amount));
    const V_proof = (0, range_1.verifyRangeProof)(proof.range, bit_size, prefix);
    if (V_proof == false) {
        throw new Error("erro in range for V");
    }
    if (!V.equals(V_proof)) {
        throw new Error("V missmatch");
    }
    ;
    let sameEncryptInputs = {
        L1: L0,
        R1: R0,
        L2: V,
        R2: R_aux,
        g: constants_1.GENERATOR,
        y1: inputs.y,
        y2: constants_1.SECONDARY_GENERATOR,
    };
    let sameEncrpyProof = {
        Ax: proof.A_x,
        AL1: proof.A,
        AL2: proof.A_v,
        AR2: proof.A_r,
        c,
        sb: proof.sb,
        sx: proof.sx,
        sr2: proof.sr,
    };
    let res = protocols_1.SameEncryptUnknownRandom.verify(sameEncryptInputs, sameEncrpyProof);
    if (res == false) {
        throw new Error("error in SameEncrpyUnkownRandom");
    }
}
