// starkzap/dist/src/bridge/operator/BridgeOperator.js
import { BridgeCache } from "../../bridge/operator/BridgeCache.js";
import { BridgeToken, EthereumBridgeToken } from "../../types/bridge/bridge-token.js";
import { Protocol } from "../../types/bridge/protocol.js";
import { ConnectedEthereumWallet, ConnectedSolanaWallet, SolanaNetwork, } from "../../connect/index.js";
import { ExternalChain, SolanaBridgeToken, } from "../../types/index.js";
import { loadEthers } from "../../connect/ethersRuntime.js";
import { loadSolanaWeb3 } from "../../connect/solanaWeb3Runtime.js";
export class BridgeOperator {
    constructor(starknetWallet, bridgingConfig) {
        this.starknetWallet = starknetWallet;
        this.bridgingConfig = bridgingConfig;
        this.cache = new BridgeCache();
    }
    async deposit(recipient, amount, token, externalWallet, options) {
        const bridge = await this.bridge(token, externalWallet, this.starknetWallet);
        return bridge.deposit(recipient, amount, options);
    }
    async getDepositBalance(token, externalWallet) {
        const bridge = await this.bridge(token, externalWallet, this.starknetWallet);
        return bridge.getAvailableDepositBalance(externalWallet.address);
    }
    async getDepositFeeEstimate(token, externalWallet, options) {
        const bridge = await this.bridge(token, externalWallet, this.starknetWallet);
        return bridge.getDepositFeeEstimate(options);
    }
    async getAllowance(token, externalWallet) {
        const bridge = await this.bridge(token, externalWallet, this.starknetWallet);
        return bridge.getAllowance();
    }
    bridge(token, wallet, starknetWallet) {
        const cached = this.cache.get(token, wallet);
        if (cached)
            return cached;
        const promise = this.createBridge(token, wallet, starknetWallet);
        this.cache.set(token, wallet, promise);
        return promise;
    }
    async createBridge(token, wallet, starknetWallet) {
        if (token.chain !== wallet.chain) {
            throw new Error(`Attempting to bridge ${token.name} on ${token.chain} but external connected wallet is on chain ${wallet.chain}. Connect to a ${token.chain} wallet`);
        }
        if (token.chain === ExternalChain.ETHEREUM) {
            return await this.createEthereumBridge(token, wallet, starknetWallet);
        }
        else if (token.chain === ExternalChain.SOLANA) {
            return await this.createSolanaBridge(token, wallet, starknetWallet);
        }
        throw new Error(`Unsupported chain "${token.chain}".`);
    }
    async createEthereumBridge(token, externalWallet, starknetWallet) {
        await loadEthers("Ethereum bridge operations");
        const walletConfig = await externalWallet.toEthWalletConfig(this.bridgingConfig?.ethereumRpcUrl);
        if (token.id === "lords") {
            const { LordsBridge } = await import("../../bridge/ethereum/lords/LordsBridge.js");
            return new LordsBridge(token, walletConfig, starknetWallet);
        }
        switch (token.protocol) {
            case Protocol.CANONICAL: {
                const { CanonicalEthereumBridge } = await import("../../bridge/ethereum/canonical/CanonicalEthereumBridge.js");
                return new CanonicalEthereumBridge(token, walletConfig, starknetWallet);
            }
            case Protocol.CCTP: {
                const { CCTPBridge } = await import("../../bridge/ethereum/cctp/CCTPBridge.js");
                return new CCTPBridge(token, walletConfig, starknetWallet);
            }
            case Protocol.OFT:
            case Protocol.OFT_MIGRATED: {
                const apiKey = this.bridgingConfig?.layerZeroApiKey;
                if (!apiKey) {
                    throw new Error("OFT bridging requires a LayerZero API key. " +
                        'Set "bridging.layerZeroApiKey" in the SDK configuration.');
                }
                const { OftBridge } = await import("../../bridge/ethereum/oft/OftBridge.js");
                return new OftBridge(token, walletConfig, starknetWallet, apiKey);
            }
            default:
                throw new Error(`Unsupported protocol "${token.protocol}" for ${token.chain} chain.`);
        }
    }
    async createSolanaBridge(token, externalWallet, starknetWallet) {
        // SolanaHyperlaneBridge and @solana/web3.js are loaded lazily in
        // to avoid pulling Node.js-only transitive dependencies
        // (@hyperlane-xyz/sdk → ethereumjs-util → assert, etc.)
        // into clients that require polyfill.
        const [{ SolanaHyperlaneBridge }, solanaWeb3] = await Promise.all([
            import("../../bridge/solana/SolanaHyperlaneBridge.js"),
            loadSolanaWeb3("Solana bridge operations"),
        ]);
        const cluster = externalWallet.network === SolanaNetwork.MAINNET
            ? "mainnet-beta"
            : "testnet";
        const endpoint = this.bridgingConfig?.solanaRpcUrl ?? solanaWeb3.clusterApiUrl(cluster);
        const connection = new solanaWeb3.Connection(endpoint);
        const walletConfig = {
            address: externalWallet.address,
            provider: externalWallet.provider,
            connection,
        };
        switch (token.protocol) {
            case Protocol.HYPERLANE:
                return await SolanaHyperlaneBridge.create(token, walletConfig, starknetWallet);
            default:
                throw new Error(`Unsupported protocol "${token.protocol}" for ${token.chain} chain.`);
        }
    }
}
