// starkzap/dist/src/types/bridge/bridge-token.js
import { Protocol, } from "../../types/bridge/protocol.js";
import { ExternalChain } from "../../types/bridge/external-chain.js";
export class BridgeToken {
    constructor(params) {
        this.id = params.id;
        this.name = params.name;
        this.symbol = params.symbol;
        if (params.coingeckoId) {
            this.coingeckoId = params.coingeckoId;
        }
        this.decimals = params.decimals;
        this.address = params.address;
        this.bridgeAddress = params.l1Bridge;
        this.starknetAddress = params.starknetAddress;
        this.starknetBridge = params.starknetBridge;
    }
}
export class EthereumBridgeToken extends BridgeToken {
    constructor(params) {
        super({ ...params });
        this.chain = ExternalChain.ETHEREUM;
        this.protocol = params.protocol;
    }
}
export class SolanaBridgeToken extends BridgeToken {
    constructor(params) {
        super({ ...params });
        this.chain = ExternalChain.SOLANA;
        this.protocol = Protocol.HYPERLANE;
    }
}
