// @fatsolutions/she/src/exceptions.ts
/**
 * Base class for all protocol-related errors.
 */
export abstract class ProtocolError extends Error {
  constructor(message: string, public readonly errorCode: string) {
    super(message);
    this.name = this.constructor.name;

    // Maintains proper stack trace for where our error was thrown (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    }
  }
}

/**
 * Thrown when input parameters have invalid lengths, formats, or values.
 */
export class InvalidInputError extends ProtocolError {
  constructor(message: string) {
    super(message, 'INVALID_INPUT');
  }
}

/**
 * Thrown when input values are out of valid ranges.
 */
export class OutOfRangeError extends ProtocolError {
  constructor(message: string, public readonly value?: bigint | number, public readonly range?: string) {
    super(message, 'OUT_OF_RANGE');
  }
}

/**
 * Thrown when preconditions for a protocol operation are not met.
 */
export class PreconditionError extends ProtocolError {
  constructor(message: string) {
    super(message, 'PRECONDITION_FAILED');
  }
}

/**
 * Thrown when array lengths don't match expected requirements.
 */
export class LengthMismatchError extends InvalidInputError {
  constructor(
    expected: number,
    actual: number,
    context: string = 'arrays'
  ) {
    super(`Length mismatch: expected ${expected} but got ${actual} for ${context}`);
  }
}
