// @fatsolutions/she/src/utils.ts
import { poseidonHashMany, ProjectivePoint, utils } from "@scure/starknet";

import { CURVE_ORDER } from "./constants.js";

/**
 * Generates a cryptographically secure random bigint within the curve order.
 * @returns {bigint} A random scalar value suitable for cryptographic operations
 */
export function generateRandom(): bigint {
  const random_bytes = utils.randomPrivateKey();
  return utils.normPrivateKeyToScalar(random_bytes);
}

/**
 * Checks if a given number is within the valid curve order range.
 * @param {bigint} number - The number to validate
 * @returns {boolean} True if the number is less than CURVE_ORDER, false otherwise
 */
export function in_curve_order(number: bigint): boolean {
  if (number < CURVE_ORDER) { return true; };
  return false;
}

/**
 * Checks if a given challenge is within the valid curve order range and not zero
 * @param {bigint} challenge- The challenge to validate
 * @returns {boolean} True if the challengeis less than CURVE_ORDER and not zero, false otherwise
 */
export function validate_challenge(challenge: bigint): boolean {
    if (challenge == 0n) { return false;}
    return in_curve_order(challenge)
}

/**
 * This is used to cast a felt252 into the curve order. It is used mainly in challenge
 * computation. We could implement more complex function to avoid non-uniform distribution in
 * the challenge space (using hashes for example). The chance of falling in a felt252 that are not
 * in CURVE_ORDER are low.
 *
 *
 * Note: The diference between a felt252 and CURVE_ORDER is ~ 2**70. So the chances of hashing
 * anything and not to land inside CURVE_ORDER are ~ 2**70/2**252.
 * 
 * @param {bigint} number - The felt252 value to cast into curve order
 * @returns {bigint} The number modulo CURVE_ORDER
 */
export function reduce_modulo_order(number: bigint) {
  return number % CURVE_ORDER;
}

/**
 * Computes a challenge value using Poseidon hash from a prefix and array of commitments.
 * Used in sigma protocols for generating verifier challenges.
 * @param {bigint} prefix - The prefix value to include in the hash
 * @param {ProjectivePoint[]} commitments - Array of elliptic curve points
 * @returns {bigint} The computed challenge value within curve order
 */
export function compute_challenge(prefix: bigint, commitments: ProjectivePoint[]) {
  const arr: bigint[] = [prefix];
  commitments.forEach(commit => {
    const { x, y } = commit.toAffine();
    arr.push(x);
    arr.push(y);
  });
  return reduce_modulo_order(poseidonHashMany(arr));
}

/**
 * Computes the s (`response`, sometimes called `z`) value in sigma protocols: s = k + x * c mod CURVE_ORDER.
 * @param {bigint} k - The random nonce value
 * @param {bigint} x - The secret/witness value
 * @param {bigint} c - The challenge value
 * @returns {bigint} The computed s value for the proof
 */
export function compute_s(k: bigint, x: bigint, c: bigint): bigint {
  return (k + x * c) % CURVE_ORDER;
}


/**
 * Computes lhs - rhs (mod curve_order)
 * @param {bigint} lhs
 * @param {bigint} rhs
 * @returns {bigint} The computed value
 */
export function subtract_modulo_curve_order(lhs: bigint, rhs:bigint): bigint {
    return ((lhs - rhs) % CURVE_ORDER + CURVE_ORDER ) % CURVE_ORDER
}
