// @fatsolutions/she/src/protocols/bit.ts
import { ProjectivePoint } from "../types.js";
import { compute_challenge, compute_s, generateRandom, validate_challenge, subtract_modulo_curve_order} from "../utils.js";
import { _verify as poe_verify } from "./poe.js";

export interface BitInputs {
  V: ProjectivePoint,
  g1: ProjectivePoint,
  g2: ProjectivePoint,
}

export interface BitProof {
  A0: ProjectivePoint,
  A1: ProjectivePoint,
  c: bigint,
  c0: bigint,
  s0: bigint,
  s1: bigint,
}

export interface BitProofWithPrefix {
  A0: ProjectivePoint,
  A1: ProjectivePoint,
  prefix: bigint,
  c0: bigint,
  s0: bigint,
  s1: bigint,
}

/**
 * Verifies a bit proof with explicit challenge.
 * @param {BitInputs} inputs - The public inputs containing V, g1, g2
 * @param {BitProof} proof - The proof containing A0, A1, c, c0, s0, s1
 * @returns {boolean} True if the proof is valid, false otherwise
 */
export function verify(inputs: BitInputs, proof: BitProof): boolean {
  const { V, g1, g2 } = inputs;
  const { A0, A1, c, c0, s0, s1 } = proof;

  return _verify(V, g1, g2, A0, A1, c, c0, s0, s1);
}

/**
 * Verifies a bit proof with challenge computed from prefix.
 * @param {BitInputs} inputs - The public inputs containing V, g1, g2
 * @param {BitProofWithPrefix} proof - The proof containing A0, A1, prefix, c0, s0, s1
 * @returns {boolean} True if the proof is valid, false otherwise
 */
export function verify_with_prefix(inputs: BitInputs, proof: BitProofWithPrefix): boolean {
  const { V, g1, g2 } = inputs;
  const { A0, A1, prefix, c0, s0, s1 } = proof;

  const commitments = [V, A0, A1];
  const c = compute_challenge(prefix, commitments);

  return _verify(V, g1, g2, A0, A1, c, c0, s0, s1);
}

/**
 * Proof of Bit: verifies that a commited `V = g^b h^r` is the ciphertext of either `b=0 OR b=1`.
 * If `b=0` then `V = h^r` and a proof of exponet for `r` is enough. If `b=1` then `V/g = h^r`
 * can be also proven with a POE. This is combined in a `OR` statement and the protocol can
 * validate that one of the cases is valid without leaking which one is the valid one.
 *
 * EC_MUL: 4;
 * EC_ADD: 3
 *
 * @param {ProjectivePoint} V - The commitment (g1^b * g2^r)
 * @param {ProjectivePoint} g1 - First generator point (for bit value)
 * @param {ProjectivePoint} g2 - Second generator point (for randomness)
 * @param {ProjectivePoint} A0 - Commitment for bit=0 case
 * @param {ProjectivePoint} A1 - Commitment for bit=1 case
 * @param {bigint} c - Challenge value
 * @param {bigint} c0 - Sub-challenge for bit=0 case
 * @param {bigint} s0 - Response for bit=0 case
 * @param {bigint} s1 - Response for bit=1 case
 * @returns {boolean} True if verification succeeds for either bit=0 or bit=1
 */
export function _verify(
  V: ProjectivePoint,
  g1: ProjectivePoint,
  g2: ProjectivePoint,
  A0: ProjectivePoint,
  A1: ProjectivePoint,
  c: bigint,
  c0: bigint,
  s0: bigint,
  s1: bigint,
): boolean {

  if (!validate_challenge(c)) { throw new Error ("invalid challenge"); };
  const c1 = subtract_modulo_curve_order(c, c0);

  if (!poe_verify(V, g2, A0, c0, s0)) {
    return false;
  }

  const V1: ProjectivePoint = V.subtract(g1);

  if (!poe_verify(V1, g2, A1, c1, s1)) {
    return false;
  }

  return true;
}


/**
 * Generates a valid transcript for a proof of exponent protocol with statement `y` and base `gen`,
 * simulating the challenge `c` and the response `s` and then deducing the correct commitment `A`
 * for the verifier.
 *
 * @param {ProjectivePoint} y - Statement to prove. Note that we actually don't know `x` such that `y = gen^x`
 * @param {ProjectivePoint} gen - Generator point
 */
function simulatePOE(
  y: ProjectivePoint,
  gen: ProjectivePoint,
) {
  const s = generateRandom();
  const c = generateRandom();
  const A = gen.multiplyUnsafe(s).subtract(y.multiplyUnsafe(c));
  return { A, c, s };
}

function _proveBit0(random: bigint, g1: ProjectivePoint, g2: ProjectivePoint, prefix: bigint): { inputs: BitInputs, proof: BitProofWithPrefix; } {
  const V = g2.multiplyUnsafe(random);
  const inputs: BitInputs = { V, g1, g2 };
  const V1 = V.subtract(g1);
  const { A: A1, c: c1, s: s1 } = simulatePOE(V1, g2);

  const k = generateRandom();
  const A0 = g2.multiply(k);

  const commitments = [V, A0, A1];
  const c = compute_challenge(prefix, commitments);

  const c0 = subtract_modulo_curve_order(c, c1);
  const s0 = compute_s(k, random, c0);

  const proof: BitProofWithPrefix = { A0, A1, prefix, c0, s0, s1 };
  return { inputs, proof };
}

function _proveBit1(random: bigint, g1: ProjectivePoint, g2: ProjectivePoint, prefix: bigint): { inputs: BitInputs, proof: BitProofWithPrefix; } {
  const V = g1.add(g2.multiplyUnsafe(random));
  const inputs: BitInputs = { V, g1, g2 };
  const V0 = V;
  const { A: A0, c: c0, s: s0 } = simulatePOE(V0, g2);

  const k = generateRandom();
  const A1 = g2.multiply(k);

  const commitments = [V, A0, A1];
  const c = compute_challenge(prefix, commitments);

  const c1 = subtract_modulo_curve_order(c, c0);
  const s1 = compute_s(k, random, c1);

  const proof: BitProofWithPrefix = { A0, A1, prefix, c0, s0, s1 };
  return { inputs, proof };
}

/**
 * Proof of Bit: generates a proof that `V = g^b h^r` is the ciphertext of either `b=0 OR b=1`.
 * If `b=0` then `V = h^r` and a proof of exponet for `r` is enough. If `b=1` then `V/g = h^r`
 * can be also proven with a POE. This is combined in a `OR` statement and the protocol can
 * validate that one of the cases is valid without leaking which one is the valid one.
 *
 * @param {0 | 1} bit - The bit value being proven (0 or 1)
 * @param {bigint} random - The randomness used in the commitment
 * @param {ProjectivePoint} g1 - First generator point (for bit value)
 * @param {ProjectivePoint} g2 - Second generator point (for randomness)
 * @param {bigint} prefix - Prefix for challenge computation
 * @returns {{inputs: BitInputs, proof: BitProofWithPrefix}} The public inputs and generated proof
 */
export function prove(bit: 0 | 1, random: bigint, g1: ProjectivePoint, g2: ProjectivePoint, prefix: bigint): { inputs: BitInputs, proof: BitProofWithPrefix; } {
  return bit === 0 ? _proveBit0(random, g1, g2, prefix) : _proveBit1(random, g1, g2, prefix);
}
