// @fatsolutions/she/src/protocols/ElGamal.ts
import { ProjectivePoint } from "../types.js";
import { compute_challenge, compute_s, generateRandom } from "../utils.js";
import { _verify as _poe_verify } from "./poe.js";
import { _verify as _poe2_verify } from "./poe2.js";

export interface ElGamalInputs {
  L: ProjectivePoint,
  R: ProjectivePoint,
  g1: ProjectivePoint,
  g2: ProjectivePoint,
}

export interface ElGamalProof {
  AL: ProjectivePoint,
  AR: ProjectivePoint,
  c: bigint,
  sb: bigint,
  sr: bigint,
}

export interface ElGamalProofWithPrefix {
  AL: ProjectivePoint,
  AR: ProjectivePoint,
  prefix: bigint,
  sb: bigint,
  sr: bigint,
}

/**
 * Verifies an ElGamal proof with explicit challenge.
 * @param {ElGamalInputs} inputs - The public inputs containing L, R, g1, g2
 * @param {ElGamalProof} proof - The proof containing AL, AR, c, sb, sr
 * @returns {boolean} True if the proof is valid, false otherwise
 */
export function verify(inputs: ElGamalInputs, proof: ElGamalProof): boolean {
  const { L, R, g1, g2 } = inputs;
  const { AL, AR, c, sb, sr } = proof;

  return _verify(L, R, g1, g2, AL, AR, c, sb, sr);
}


/**
 * Verifies an ElGamal proof with challenge computed from prefix.
 * @param {ElGamalInputs} inputs - The public inputs containing L, R, g1, g2
 * @param {ElGamalProofWithPrefix} proof - The proof containing AL, AR, prefix, sb, sr
 * @returns {boolean} True if the proof is valid, false otherwise
 */
export function verify_with_prefix(inputs: ElGamalInputs, proof: ElGamalProofWithPrefix): boolean {
  const { L, R, g1, g2 } = inputs;
  const { AL, AR, prefix, sb, sr } = proof;
  const commitments = [AL, AR];
  const c = compute_challenge(prefix, commitments);
  return _verify(L, R, g1, g2, AL, AR, c, sb, sr);
}

/**
 * Verifies that a pair of EC points is a well formed ElGamal encryption of the form
 * `(L, R) = (g1^b g2^r , g1^r)`
 *
 * @param {ProjectivePoint} L - Left component of ElGamal ciphertext (g1^b * g2^r)
 * @param {ProjectivePoint} R - Right component of ElGamal ciphertext (g1^r)
 * @param {ProjectivePoint} g1 - First generator point
 * @param {ProjectivePoint} g2 - Second generator point
 * @param {ProjectivePoint} AL - Left commitment (g1^kb * g2^kr)
 * @param {ProjectivePoint} AR - Right commitment (g1^kr)
 * @param {bigint} c - Challenge value
 * @param {bigint} sb - Response for message (kb + c*b)
 * @param {bigint} sr - Response for randomness (kr + c*r)
 * @returns {boolean} True if verification succeeds, false if POE or POE2 verification fails
 */
export function _verify(
  L: ProjectivePoint,
  R: ProjectivePoint,
  g1: ProjectivePoint,
  g2: ProjectivePoint,
  AL: ProjectivePoint,
  AR: ProjectivePoint,
  c: bigint,
  sb: bigint,
  sr: bigint,
): boolean {

  if (!_poe_verify(R, g1, AR, c, sr)) {
    return false;
  }

  if (!_poe2_verify(L, g1, g2, AL, c, sb, sr)) {
    return false;
  }
  return true;
}


/**
 * Generates an ElGamal ciphertext of form `(L, R)` that encrypts the given
 * message `b` and a zero-knowledge proof that `(L, R) = (g1^b g2^r , g1^r)`
 *
 * The sigma protocol consists of a coupled POE and POE2 and runs as
 * follows:
 *
 * ```
 *     P:  kb,kr <-- R        sends    AL = g1^kb g2^kr, AR=g1^kr
 *     V:  c <-- R            sends    c
 *     P:  sb = kb + c*b
 *     P:  sr = kr + c*r      sends s1, s1
 * ```
 *
 * The verifier asserts:
 *
 * - `g1^sr       == AR * (R^c)   [poe]`
 *
 * - `g1^sb g2^sr == AL * (L^c)   [poe2]`
 *
 * notation: (P)rover (V)erifier (R)andom
 *
 * @param {bigint} message - The message being encrypted (b)
 * @param {bigint} random - The randomness used in encryption (r)
 * @param {ProjectivePoint} g1 - First generator point
 * @param {ProjectivePoint} g2 - Second generator point
 * @param {bigint} prefix - Prefix for challenge computation
 * @returns {{inputs: ElGamalInputs, proof: ElGamalProofWithPrefix}} The public inputs and generated proof
 */
export function prove(
  message: bigint,
  random: bigint,
  g1: ProjectivePoint,
  g2: ProjectivePoint,
  prefix: bigint,
): { inputs: ElGamalInputs, proof: ElGamalProofWithPrefix; } {

  const L = g1.multiply(message).add(g2.multiplyUnsafe(random));
  const R = g1.multiplyUnsafe(random);

  const kb = generateRandom();
  const kr = generateRandom();

  const AL = g1.multiply(kb).add(g2.multiply(kr));
  const AR = g1.multiply(kr);

  const c = compute_challenge(prefix, [AL, AR]);

  const sr = compute_s(kr, random, c);
  const sb = compute_s(kb, message, c);

  const inputs: ElGamalInputs = { L, R, g1, g2 };
  const proof: ElGamalProofWithPrefix = { AL, AR, sb, sr, prefix };
  return { inputs, proof };
}
