// @fatsolutions/she/src/protocols/poe.ts
import { ProjectivePoint } from "../types.js";
import { compute_challenge, compute_s, generateRandom, in_curve_order, validate_challenge } from "../utils.js";

export interface PoeInputs {
  y: ProjectivePoint,
  g: ProjectivePoint,
}

export interface PoeProof {
  A: ProjectivePoint,
  c: bigint,
  s: bigint,
}

export interface PoeProofWithPrefix {
  A: ProjectivePoint,
  prefix: bigint,
  s: bigint,
}

/**
 * Verifies a Proof of Exponent (POE) with explicit challenge.
 * @param {PoeInputs} inputs - The public inputs containing y and g
 * @param {PoeProof} proof - The proof containing A, c, and s
 * @returns {boolean} True if the proof is valid, false otherwise
 */
export function verify(inputs: PoeInputs, proof: PoeProof): boolean {
  const { y, g } = inputs;
  const { A, c, s } = proof;
  return _verify(y, g, A, c, s);
}

/**
 * Verifies a Proof of Exponent (POE) with challenge computed from prefix.
 * @param {PoeInputs} inputs - The public inputs containing y and g
 * @param {PoeProofWithPrefix} proof - The proof containing A, prefix, and s
 * @returns {boolean} True if the proof is valid, false otherwise
 */
export function verify_with_prefix(inputs: PoeInputs, proof: PoeProofWithPrefix): boolean {
  const { y, g } = inputs;
  const { A, prefix, s } = proof;
  const commitments = [A];
  const c = compute_challenge(prefix, commitments);

  return _verify(y, g, A, c, s);
}


/**
 * Proof of Exponent: validate a proof of knowledge of the exponent `x` such that `y = g ^ x`.
 * The sigma protocol runs as follow:
 *
 * ```
 *     P:  k <-- R        sends    A = g ^ k
 *     V:  c <-- R        sends    c
 *     P:  s = k + c*x    sends    s
 * ```
 *
 * The verifier asserts:
 *
 * - `g^s == A * (y^c)`
 *
 * notation: (P)rover (V)erifier (R)andom
 *
 * EC_MUL: 2;
 * EC_ADD: 1
 *
 * @param {ProjectivePoint} y - The public key (g^x)
 * @param {ProjectivePoint} g - The generator point
 * @param {ProjectivePoint} A - The commitment (g^k)
 * @param {bigint} c - The challenge value
 * @param {bigint} s - The response value (k + c*x)
 * @returns {boolean} True if the proof verification succeeds
 */
export function _verify(
  y: ProjectivePoint,
  g: ProjectivePoint,
  A: ProjectivePoint,
  c: bigint,
  s: bigint,
): boolean {
  if (!in_curve_order(s)) { throw new Error(" s not in curve order"); };
  if (!validate_challenge(c)) { throw new Error ("invalid challenge"); };
  const LHS = g.multiplyUnsafe(s);
  const RHS = A.add(y.multiplyUnsafe(c));
  return LHS.equals(RHS);
}


/**
 * Generates a Proof of Exponent (POE) for knowledge of discrete logarithm x where y = g^x.
 *
 * @param {bigint} x - The secret exponent/private key
 * @param {ProjectivePoint} g - The generator point
 * @param {bigint} prefix - The prefix for challenge computation
 * @returns {{inputs: PoeInputs, proof: PoeProofWithPrefix}} The public inputs and generated proof
 */
export function prove(x: bigint, g: ProjectivePoint, prefix: bigint): { inputs: PoeInputs, proof: PoeProofWithPrefix; } {
  const y = g.multiply(x);
  const inputs: PoeInputs = { y, g };
  const k = generateRandom();
  const A = g.multiplyUnsafe(k);
  const c = compute_challenge(prefix, [A]);
  const s = compute_s(k, x, c);
  const proof: PoeProofWithPrefix = { A, s, prefix };
  return { inputs, proof };
}
