// @fatsolutions/she/src/protocols/poe2.ts
import { ProjectivePoint } from "../types.js";
import { compute_challenge, compute_s, generateRandom, validate_challenge, in_curve_order } from "../utils.js";

export interface Poe2Inputs {
  y: ProjectivePoint,
  g1: ProjectivePoint,
  g2: ProjectivePoint,
}

export interface Poe2Proof {
  A: ProjectivePoint,
  c: bigint,
  s1: bigint,
  s2: bigint,
}

export interface Poe2ProofWithPrefix {
  A: ProjectivePoint,
  prefix: bigint,
  s1: bigint,
  s2: bigint,
}

/**
 * Verifies a Proof of Exponent 2 (POE2) with explicit challenge.
 * @param {Poe2Inputs} inputs - The public inputs containing y, g1, g2
 * @param {Poe2Proof} proof - The proof containing A, c, s1, s2
 * @returns {boolean} True if the proof is valid, false otherwise
 */
export function verify(inputs: Poe2Inputs, proof: Poe2Proof): boolean {
  const { y, g1, g2 } = inputs;
  const { A, c, s1, s2 } = proof;
  return _verify(y, g1, g2, A, c, s1, s2);
}

/**
 * Verifies a Proof of Exponent 2 (POE2) with challenge computed from prefix.
 * @param {Poe2Inputs} inputs - The public inputs containing y, g1, g2
 * @param {Poe2ProofWithPrefix} proof - The proof containing A, prefix, s1, s2
 * @returns {boolean} True if the proof is valid, false otherwise
 */
export function verify_with_prefix(inputs: Poe2Inputs, proof: Poe2ProofWithPrefix): boolean {
  const { y, g1, g2 } = inputs;
  const { A, prefix, s1, s2 } = proof;
  const commitments = [A];
  const c = compute_challenge(prefix, commitments);

  return _verify(y, g1, g2, A, c, s1, s2);
}


/**
 * Proof of Exponent 2: validate a proof of knowledge for the exponents `y = g1^x1 g2^x2`,
 * where `x1` and `x2` are secret and the bases `g1, g2` are public. The sigma
 * protocol (formally known as Okamoto's protocol) runs as follows:
 *
 * ```
 *     P:  k1,k2 <-- R        sends    A = g1^k1 g2^k2
 *     V:  c <-- R            sends    c
 *     P:  s1 = k1 + c*x1
 *     P:  s2 = k2 + c*x2     sends    s1, s2
 * ```
 *
 * The verifier asserts:
 *
 * - `g1^s1 g2^s2 == A * (y^c)`
 *
 * notation: (P)rover (V)erifier (R)andom
 *
 * EC_MUL: 3;
 * EC_ADD: 2
 * 
 * @param {ProjectivePoint} y - The public commitment (g1^x1 * g2^x2)
 * @param {ProjectivePoint} g1 - First generator point
 * @param {ProjectivePoint} g2 - Second generator point
 * @param {ProjectivePoint} A - The commitment (g1^k1 * g2^k2)
 * @param {bigint} c - The challenge value
 * @param {bigint} s1 - First response value (k1 + c*x1)
 * @param {bigint} s2 - Second response value (k2 + c*x2)
 * @returns {boolean} True if the proof verification succeeds
 */
export function _verify(
  y: ProjectivePoint,
  g1: ProjectivePoint,
  g2: ProjectivePoint,
  A: ProjectivePoint,
  c: bigint,
  s1: bigint,
  s2: bigint,
): boolean {
  if (!in_curve_order(s1)) { throw new Error(" s1 not in curve order"); };
  if (!in_curve_order(s2)) { throw new Error(" s2 not in curve order"); };
  if (!validate_challenge(c)) { throw new Error ("invalid challenge"); };

  const LHS = g1.multiplyUnsafe(s1).add(g2.multiplyUnsafe(s2));
  const RHS = A.add(y.multiplyUnsafe(c));
  return LHS.equals(RHS);
}


/**
 * Generates a Proof of Exponent 2 (POE2) for knowledge of discrete logarithms `x1`, `x2`
 * such that `y = g1^x1 * g2^x2`, and `x1, x2` are secret and the bases `g1, g2` are public.
 * The sigma protocol (formally known as Okamoto's protocol) runs as follows:
 *
 * ```
 *     P:  k1,k2 <-- R        sends    A = g1^k1 g2^k2
 *     V:  c <-- R            sends    c
 *     P:  s1 = k1 + c*x1
 *     P:  s2 = k2 + c*x2     sends    s1, s2
 * ```
 *
 * The verifier asserts:
 *
 * - `g1^s1 g2^s2 == A * (y^c)`
 *
 * notation: (P)rover (V)erifier (R)andom
 *
 * @param {bigint} x1 - The first secret exponent
 * @param {bigint} x2 - The second secret exponent
 * @param {ProjectivePoint} g1 - First generator point
 * @param {ProjectivePoint} g2 - Second generator point
 * @param {bigint} prefix - Prefix for challenge computation
 * @returns {{inputs: Poe2Inputs, proof: Poe2ProofWithPrefix}} The public inputs and generated proof
 */
export function prove(x1: bigint, x2: bigint, g1: ProjectivePoint, g2: ProjectivePoint, prefix: bigint): { inputs: Poe2Inputs, proof: Poe2ProofWithPrefix; } {
  const y = g1.multiply(x1).add(g2.multiply(x2));
  const inputs: Poe2Inputs = { y, g1, g2 };
  const k1 = generateRandom();
  const k2 = generateRandom();
  const A = g1.multiplyUnsafe(k1).add(g2.multiplyUnsafe(k2));
  const c = compute_challenge(prefix, [A]);
  const s1 = compute_s(k1, x1, c);
  const s2 = compute_s(k2, x2, c);
  const proof: Poe2ProofWithPrefix = { A, prefix, s1, s2 };
  return { inputs, proof };
}
