// @fatsolutions/she/src/protocols/poeN.ts
import { LengthMismatchError } from "../exceptions.js";
import { ProjectivePoint } from "../types.js";
import { compute_challenge, compute_s, generateRandom, in_curve_order, validate_challenge} from "../utils.js";


export interface PoeNInputs {
  y: ProjectivePoint,
  generators: ProjectivePoint[],
}

export interface PoeNProof {
  A: ProjectivePoint,
  c: bigint,
  ss: bigint[],
}

export interface PoeNProofWithPrefix {
  A: ProjectivePoint,
  prefix: bigint,
  ss: bigint[],
}

/**
 * Verifies a Proof of Exponents (POEN) with explicit challenge.
 * @param {PoeNInputs} inputs - The public inputs containing y and generators
 * @param {PoeNProof} proof - The proof containing A, c, and ss
 * @returns {boolean} True if the proof is valid, false otherwise
 */
export function verify(inputs: PoeNInputs, proof: PoeNProof): boolean {
  const { y, generators } = inputs;
  const { A, c, ss } = proof;
  return _verify(y, generators, A, c, ss);
}

/**
 * Verifies a Proof of Exponents (POEN) with challenge computed from prefix.
 * @param {PoeNInputs} inputs - The public inputs containing y and generators
 * @param {PoeNProofWithPrefix} proof - The proof containing A, prefix, and ss
 * @returns {boolean} True if the proof is valid, false otherwise
 */
export function verify_with_prefix(inputs: PoeNInputs, proof: PoeNProofWithPrefix): boolean {
  const { y, generators } = inputs;
  const { A, prefix, ss } = proof;
  const commitments = [A];
  const c = compute_challenge(prefix, commitments);

  return _verify(y, generators, A, c, ss);
}

/**
 * Proof of Exponents: validate a proof of knowledge of the exponents `y = g_1^x_1 g_2^x_2 ... g_N^x_N`.
 * The sigma protocol runs as follows:
 *
 * ```
 *     P:  k_i <-- R          sends    A = g_1^k_1... g_i^k_i ... g_N^k_N
 *     V:  c <-- R            sends    c
 *     P:  s_i = k_i + c*x_i  sends    s_i
 *     The verifier asserts:
 *     - g_1^s_1...g_i^s_i... g_N^s_N == A * (y^c)
 * ```
 *
 * notation: (P)rover (V)erifier (R)andom
 *
 * EC_MUL: N+1;
 * EC_ADD: N
 * 
 * @param {ProjectivePoint} y - The public commitment (g_1^x_1 * g_2^x_2 * ... * g_N^x_N)
 * @param {ProjectivePoint[]} generators - Array of generator points [g_1, g_2, ..., g_N]
 * @param {ProjectivePoint} A - The commitment (g_1^k_1 * g_2^k_2 * ... * g_N^k_N)
 * @param {bigint} c - The challenge value
 * @param {bigint[]} ss - Array of response values [s_1, s_2, ..., s_N]
 * @returns {boolean} True if the proof verification succeeds
 * @throws {LengthMismatchError} If generators and responses arrays have different lengths
 */
export function _verify(
  y: ProjectivePoint,
  generators: ProjectivePoint[],
  A: ProjectivePoint,
  c: bigint,
  ss: bigint[],
): boolean {
  if (generators.length !== ss.length) {
    throw new LengthMismatchError(generators.length, ss.length, 'generators and responses');
  }
  if (!validate_challenge(c)) { throw new Error ("invalid challenge"); };

  let LHS = ProjectivePoint.ZERO;
  for (let i = 0; i < generators.length; i++) {
    if (!in_curve_order(ss[i]!)) { throw new Error(" s not in curve order"); };
    LHS = LHS.add(generators[i]!.multiplyUnsafe(ss[i]!));
  }

  const RHS = A.add(y.multiplyUnsafe(c));

  return LHS.equals(RHS);
}


/**
 * Proof of Exponents: generates a proof of knowledge for the exponents `x_i` in `y = g_1^x_1 g_2^x_2 ... g_N^x_N`.
 * The sigma protocol (a generalization of Okamoto's) runs as follows:
 *
 * ```
 *     P:  k_i <-- R          sends    A = g_1^k_1... g_i^k_i ... g_N^k_N
 *     V:  c <-- R            sends    c
 *     P:  s_i = k_i + c*x_i  sends    s_i
 *     The verifier asserts:
 *     - g_1^s_1...g_i^s_i... g_N^s_N == A * (y^c)
 * ```
 *
 * notation: (P)rover (V)erifier (R)andom
 *
 * @param {bigint[]} scalars - Array of secret exponents [x_1, x_2, ..., x_N]
 * @param {ProjectivePoint[]} generators - Array of generator points [g_1, g_2, ..., g_N]
 * @param {bigint} prefix - Prefix for challenge computation
 * @returns {{inputs: PoeNInputs, proof: PoeNProofWithPrefix}} The public inputs and generated proof
 * @throws {LengthMismatchError} If scalars and generators arrays have different lengths
 */
export function prove(
  scalars: bigint[],
  generators: ProjectivePoint[],
  prefix: bigint,
): { inputs: PoeNInputs, proof: PoeNProofWithPrefix; } {

  if (generators.length !== scalars.length) {
    throw new LengthMismatchError(generators.length, scalars.length, 'generators and scalars');
  }

  let ks: bigint[] = [];
  let y = ProjectivePoint.ZERO;
  let A = ProjectivePoint.ZERO;
  for (let i = 0; i < generators.length; i++) {
    const g = generators[i]!;
    const k = generateRandom();

    y = y.add(g.multiply(scalars[i]!));
    A = A.add(g.multiply(k));
    ks.push(k);
  }

  const inputs: PoeNInputs = { y, generators };

  const commitments = [A];
  const c = compute_challenge(prefix, commitments);

  const ss: bigint[] = [];
  for (let i = 0; i < generators.length; i++) {
    ss.push(compute_s(ks[i]!, scalars[i]!, c));
  }

  const proof: PoeNProofWithPrefix = { A, prefix, ss };

  return { inputs, proof };
}
