// @fatsolutions/she/src/protocols/range.ts
import { CURVE_ORDER } from "../constants.js";
import { LengthMismatchError, OutOfRangeError } from "../exceptions.js";
import { ProjectivePoint } from "../types.js";
import { generateRandom } from "../utils.js";
import {
  BitInputs,
  BitProofWithPrefix,
  prove as bit_prove,
  verify_with_prefix as bit_verify
} from "./bit.js";


export interface RangeInputs {
  g1: ProjectivePoint,
  g2: ProjectivePoint,
  bit_size: number,
  commitments: ProjectivePoint[],
}

export interface RangeProof {
  proofs: BitProofWithPrefix[],
}

/**
 * Verify that a span of `V_i = g^b_i h^r_i` are encoding either `b=1` or `b=0` and that
 * those `b_i` are indeed the binary decomposition `b = sum_i b_i 2^i`. With the `b` that
 * is encoded in `V = g^b h^r`. (Note that `r = sum_i r_i 2^i`)
 *
 *
 * EC_MUL: bit_size * ( 4 + 1 ) (160 for u32);
 * EC_ADD: bit_size * ( 3 + 1 ) (128 for u32)
 * 
 * @param {RangeInputs} inputs - The range proof inputs containing g1, g2, bit_size, and commitments
 * @param {RangeProof} proof - The range proof containing bit proofs for each commitment
 * @returns {false | ProjectivePoint} Computed V, if all bit proofs verify and commitments are consistent, false otherwise
 * @throws {LengthMismatchError} If input lengths mismatch
 */
export function verify(inputs: RangeInputs, proof: RangeProof): false | ProjectivePoint {
  const { g1, g2, bit_size, commitments } = inputs;
  const { proofs } = proof;

  if (commitments.length !== bit_size || proofs.length !== bit_size) {
    throw new LengthMismatchError(commitments.length, bit_size, 'commitments and bit_size');
  }

  const initial_prefix = proofs[0]!.prefix;
  let V_total = ProjectivePoint.ZERO;
  for (let i = 0; i < bit_size; i++) {
    const pow = 2n ** BigInt(i);
    const V = commitments[i]!;  // we just checked length
    const pi = proofs[i]!;
    if (pi.prefix !== initial_prefix + BigInt(i)) {
      return false;
    }

    const bit_inputs: BitInputs = { V, g1, g2 };
    if (!bit_verify(bit_inputs, pi)) {
      return false;
    }

    V_total = V_total.add(V.multiply(pow));
  }
  return V_total;
};

export function pregenerate_randomness(bit_size:number): {randomness:bigint[], total_random: bigint}  {
  let randomness: bigint[] = [];
  let r = 0n;
  for (let i = 0; i < bit_size; i++) {
    const pow = 2n ** BigInt(i);
    const r_i = generateRandom();
    randomness.push(r_i);
    r = (r + r_i * pow) % CURVE_ORDER;
  }
  return {randomness, total_random:r}
}

/**
 * Generates a range proof that a value b is within [0, 2^bit_size - 1].
 * Creates bit commitments for the binary decomposition of b and proves each bit is 0 or 1.
 *
 * @param {bigint} b - The value to prove is in range
 * @param {number} bit_size - The number of bits (range is [0, 2^bit_size - 1])
 * @param {ProjectivePoint} g1 - First generator point
 * @param {ProjectivePoint} g2 - Second generator point
 * @param {bigint} prefix - Prefix for challenge computation
 * @returns {{inputs: RangeInputs, proofs: RangeProof, r: bigint}} The inputs, proofs, and total randomness
 * @throws {OutOfRangeError} If b is not in the specified range
 */
export function prove(
  b: bigint,
  bit_size: number,
  g1: ProjectivePoint,
  g2: ProjectivePoint,
  randomness:bigint[],
  initial_prefix: bigint,
): { inputs: RangeInputs, proofs: RangeProof, r: bigint; } {
  if (b >= 2 ** bit_size) {
    throw new OutOfRangeError(`Value ${b} is not in range [0, ${2 ** bit_size - 1}]`, b, `[0, ${2 ** bit_size - 1}]`);
  }
  if (randomness.length !== bit_size) {
    throw new Error(`Randomness length does not match bit_size`)
  }
  const b_bin: (0 | 1)[] = b
    .toString(2)
    .padStart(bit_size, "0")
    .split("")
    .map(Number)
    .map(x => x as (0 | 1))
    .reverse();

  const proofs: BitProofWithPrefix[] = [];
  const commitments: ProjectivePoint[] = [];
  let r = 0n;
  for (let i = 0; i < bit_size; i++) {
    const pow = 2n ** BigInt(i);
    const r_inn = randomness[i]!
    const { inputs, proof } = bit_prove(b_bin[i]!, r_inn, g1, g2, initial_prefix + BigInt(i));
    proofs.push(proof);
    commitments.push(inputs.V);
    r = (r + r_inn * pow) % CURVE_ORDER;
  }
  const inputs: RangeInputs = { commitments, g1, g2, bit_size };
  const rangeproof: RangeProof = { proofs };
  return { inputs, proofs: rangeproof, r };
}
