// @fatsolutions/she/src/protocols/sameEncryptionUnknownRandom.ts
import { ProjectivePoint } from "../types.js";
import { compute_challenge, compute_s, generateRandom } from "../utils.js";
import { PreconditionError } from "../exceptions.js";
import { _verify as _elgamal_verify } from "./ElGamal.js";
import { _verify as _poe_verify } from "./poe.js";
import { _verify as _poe2_verify } from "./poe2.js";

export interface SameEncryptionUnknownRandomInputs {
  L1: ProjectivePoint,
  R1: ProjectivePoint,
  L2: ProjectivePoint,
  R2: ProjectivePoint,
  g: ProjectivePoint,
  y1: ProjectivePoint,
  y2: ProjectivePoint,
}

export interface SameEncryptionUnknownRandomProof {
  Ax: ProjectivePoint,
  AL1: ProjectivePoint,
  AL2: ProjectivePoint,
  AR2: ProjectivePoint,
  c: bigint,
  sb: bigint,
  sx: bigint,
  sr2: bigint,
}

export interface SameEncryptionUnknownRandomProofWithPrefix {
  Ax: ProjectivePoint,
  AL1: ProjectivePoint,
  AL2: ProjectivePoint,
  AR2: ProjectivePoint,
  prefix: bigint,
  sb: bigint,
  sx: bigint,
  sr2: bigint,
}

/**
 * Verifies a same encryption unknown random proof with explicit challenge.
 * @param {SameEncryptionUnknownRandomInputs} inputs - The public inputs containing L1, R1, L2, R2, g, y1, y2
 * @param {SameEncryptionUnknownRandomProof} proof - The proof containing Ax, AL1, AL2, AR2, c, sb, sx, sr2
 * @returns {boolean} True if the proof is valid, false otherwise
 */
export function verify(
  inputs: SameEncryptionUnknownRandomInputs, proof: SameEncryptionUnknownRandomProof,
): boolean {
  const { L1, R1, L2, R2, g, y1, y2 } = inputs;
  const { Ax, AL1, AL2, AR2, c, sb, sx, sr2 } = proof;

  return _verify(L1, R1, L2, R2, g, y1, y2, Ax, AL1, AL2, AR2, c, sb, sx, sr2);
}

/**
 * Verifies a same encryption unknown random proof with challenge computed from prefix.
 * @param {SameEncryptionUnknownRandomInputs} inputs - The public inputs containing L1, R1, L2, R2, g, y1, y2
 * @param {SameEncryptionUnknownRandomProofWithPrefix} proof - The proof containing Ax, AL1, AL2, AR2, prefix, sb, sx, sr2
 * @returns {boolean} True if the proof is valid, false otherwise
 */
export function verify_with_prefix(
  inputs: SameEncryptionUnknownRandomInputs, proof: SameEncryptionUnknownRandomProofWithPrefix,
): boolean {
  const { L1, R1, L2, R2, g, y1, y2 } = inputs;
  const { Ax, AL1, AL2, AR2, prefix, sb, sx, sr2 } = proof;
  const commitments = [Ax, AL1, AL2, AR2];
  const c = compute_challenge(prefix, commitments);

  return _verify(L1, R1, L2, R2, g, y1, y2, Ax, AL1, AL2, AR2, c, sb, sx, sr2);
}

/**
 * Verifies that two encryptions for two keys are valid and that they are encrypting the same
 * amount `b`. For this proof, the prover knows only one of the randoms values and knows the secret
 * of the public key that does not know the random for.
 * Let `(L1, R1) = (g^b y^_r, g^_r)`, `_r` is unknown to the prover. The prover can decrypt `b`
 * with the knowledge of `x`.
 * By showing that `L1 = g^b R^x` whe can prove that the ElGamal encryption indeed encrypts `b`.
 * The protocol runs as follows
 *
 * ```
 *     P:  kx, kb, kr <-- R     sends    Ax=g^kx, AL1 = g^kb R^kx, AL2 = g^kb y2^kr, AR2 = g^kr
 *     V:  c <-- R              sends    c
 *     P:  sx = k + c*x         sends    s
 *     P:  sb = kb + c*b        sends    sb
 *     P:  sr = kr + c*r        sends    sr
 * ```
 *
 * The verifier asserts:
 *
 * - `g^sx == Ax * (y^c)          [poe]`
 *
 * - `g^sb R1^sx == AL1 * (L1^c)  [poe2]`
 *
 * - `verifyElGamal for (L2,R2)`
 *
 * notation: (P)rover (V)erifier (R)andom
 *
 * EC_MUL: 10;
 * EC_ADD: 6
 *
 * @param {ProjectivePoint} L1 - Left component of first ElGamal ciphertext (unknown random)
 * @param {ProjectivePoint} R1 - Right component of first ElGamal ciphertext (unknown random)
 * @param {ProjectivePoint} L2 - Left component of second ElGamal ciphertext (known random)
 * @param {ProjectivePoint} R2 - Right component of second ElGamal ciphertext (known random)
 * @param {ProjectivePoint} g - Generator point
 * @param {ProjectivePoint} y1 - First public key (secret known to prover)
 * @param {ProjectivePoint} y2 - Second public key
 * @param {ProjectivePoint} Ax - Commitment for secret key
 * @param {ProjectivePoint} AL1 - Left commitment for first encryption
 * @param {ProjectivePoint} AL2 - Left commitment for second encryption
 * @param {ProjectivePoint} AR2 - Right commitment for second encryption
 * @param {bigint} c - Challenge value
 * @param {bigint} sb - Response for message
 * @param {bigint} sx - Response for secret key
 * @param {bigint} sr2 - Response for second randomness
 * @returns {boolean} True if verification succeeds, false if POE, POE2, or ElGamal verification fails
 */
export function _verify(
  L1: ProjectivePoint,
  R1: ProjectivePoint,
  L2: ProjectivePoint,
  R2: ProjectivePoint,
  g: ProjectivePoint,
  y1: ProjectivePoint,
  y2: ProjectivePoint,
  Ax: ProjectivePoint,
  AL1: ProjectivePoint,
  AL2: ProjectivePoint,
  AR2: ProjectivePoint,
  c: bigint,
  sb: bigint,
  sx: bigint,
  sr2: bigint,
): boolean {

  if (!_poe_verify(y1, g, Ax, c, sx)) {
    return false;
  }

  if (!_poe2_verify(L1, g, R1, AL1, c, sb, sx)) {
    return false;
  }

  if (!_elgamal_verify(L2, R2, g, y2, AL2, AR2, c, sb, sr2)) {
    return false;
  }
  return true;
}

/**
 * Generates a proof of two encryptions for two keys that are valid and that they are encrypting the
 * same amount `b`. For this proof, the prover knows only one of the randoms values and knows the
 * secret of the public key that does not know the random for. Let `(L1, R1) = (g^b y^_r, g^_r)`,
 * `_r` is unknown to the prover. The prover can decrypt `b` with the knowledge of `x`. By showing
 * that `L1 = g^b R^x` they can prove that the ElGamal encryption indeed encrypts `b`.
 *
 * The protocol runs as follows
 *
 * ```
 *     P:  kx, kb, kr <-- R     sends    Ax=g^kx, AL1 = g^kb R^kx, AL2 = g^kb y2^kr, AR2 = g^kr
 *     V:  c <-- R              sends    c
 *     P:  sx = k + c*x         sends    s
 *     P:  sb = kb + c*b        sends    sb
 *     P:  sr = kr + c*r        sends    sr
 * ```
 *
 * The verifier asserts:
 *
 * - `g^sx == Ax * (y^c)          [poe]`
 *
 * - `g^sb R1^sx == AL1 * (L1^c)  [poe2]`
 *
 * - `verifyElGamal for (L2,R2)`
 *
 * notation: (P)rover (V)erifier (R)andom
 *
 * @param {ProjectivePoint} g - Generator point
 * @param {bigint} secret1 - Secret key for first public key (y1 = g^secret1)
 * @param {ProjectivePoint} y2 - Second public key
 * @param {ProjectivePoint} L1 - Left component of first ciphertext (unknown random)
 * @param {ProjectivePoint} R1 - Right component of first ciphertext (unknown random)
 * @param {bigint} message - The message being encrypted in both ciphertexts
 * @param {bigint} random2 - Randomness for second encryption
 * @param {bigint} prefix - Prefix for challenge computation
 * @returns {{inputs: SameEncryptionUnknownRandomInputs, proof: SameEncryptionUnknownRandomProofWithPrefix}} The public inputs and generated proof
 * @throws {PreconditionError} If (L1, R1) is not a valid encryption of message under the derived public key
 */
export function prove(
  g: ProjectivePoint,
  secret1: bigint,
  y2: ProjectivePoint,
  L1: ProjectivePoint,
  R1: ProjectivePoint,
  message: bigint,
  random2: bigint,
  prefix: bigint,
): { inputs: SameEncryptionUnknownRandomInputs, proof: SameEncryptionUnknownRandomProofWithPrefix; } {
  const y = g.multiply(secret1);

  //this is to assert that (L1,R1) are a cipherbalance for message under y = g**x
  const g_message = L1.subtract(R1.multiply(secret1));
  if (!g_message.equals(g.multiply(message))) {
    throw new PreconditionError("The provided ciphertext (L1, R1) does not encrypt the specified message under the derived public key");
  }

  const L2 = g.multiply(message).add(y2.multiply(random2));
  const R2 = g.multiply(random2);

  const inputs: SameEncryptionUnknownRandomInputs = { L1, R1, L2, R2, g, y1: y, y2 };

  const kx1 = generateRandom();
  const kb = generateRandom();
  const kr2 = generateRandom();

  const Ax = g.multiply(kx1);
  const AL1 = g.multiply(kb).add(R1.multiply(kx1));

  const AL2 = g.multiply(kb).add(y2.multiply(kr2));
  const AR2 = g.multiply(kr2);

  const commitments = [Ax, AL1, AL2, AR2];
  const c = compute_challenge(prefix, commitments);

  const sx = compute_s(kx1, secret1, c);
  const sb = compute_s(kb, message, c);
  const sr2 = compute_s(kr2, random2, c);

  const proof: SameEncryptionUnknownRandomProofWithPrefix = { Ax, AL1, AL2, AR2, prefix, sx, sb, sr2 };

  return { inputs, proof };
}
