// starkzap/dist/src/swap/avnu.js
import { DEFAULT_AVNU_API_BASES, normalizeAvnuCalls, supportsAvnuChain, withAvnuApiBaseFallback, } from "../utils/avnu.js";
import { getQuotes, quoteToCalls } from "@avnu/avnu-sdk";
const DEFAULT_QUOTES_PAGE_SIZE = 5;
const DEFAULT_SLIPPAGE_BPS = 100n;
const BPS_DENOMINATOR = 10000n;
function validateQuotesPageSize(value) {
    if (!Number.isInteger(value) || value <= 0) {
        throw new Error("AVNU quotesPageSize must be a positive integer");
    }
    return value;
}
function bpsToPercent(bps) {
    if (bps < 0n || bps >= BPS_DENOMINATOR) {
        throw new Error("Invalid slippage bps");
    }
    return Number(bps) / Number(BPS_DENOMINATOR);
}
function percentToBps(value) {
    if (value == null || Number.isNaN(value)) {
        return null;
    }
    return BigInt(Math.round(value * 100));
}
function toSwapQuote(params) {
    const normalizedQuote = {
        amountInBase: params.quote.sellAmount,
        amountOutBase: params.quote.buyAmount,
        priceImpactBps: percentToBps(params.quote.priceImpact ?? null),
        provider: "avnu",
    };
    if (params.routeCallCount != null) {
        normalizedQuote.routeCallCount = params.routeCallCount;
    }
    return normalizedQuote;
}
export class AvnuSwapProvider {
    constructor(options = {}) {
        this.id = "avnu";
        this.apiBases = {
            SN_MAIN: options.apiBases?.SN_MAIN ?? [...DEFAULT_AVNU_API_BASES.SN_MAIN],
            SN_SEPOLIA: options.apiBases?.SN_SEPOLIA ?? [
                ...DEFAULT_AVNU_API_BASES.SN_SEPOLIA,
            ],
        };
        this.quotesPageSize = validateQuotesPageSize(options.quotesPageSize ?? DEFAULT_QUOTES_PAGE_SIZE);
    }
    supportsChain(chainId) {
        return supportsAvnuChain(chainId);
    }
    async getQuote(request) {
        const { quote } = await this.fetchQuoteForRequest(request);
        return toSwapQuote({ quote });
    }
    async prepareSwap(request) {
        const { quote, apiBase } = await this.fetchQuoteForRequest(request);
        const slippage = bpsToPercent(request.slippageBps ?? DEFAULT_SLIPPAGE_BPS);
        const quoteToCallsRequest = {
            quoteId: quote.quoteId,
            slippage,
            executeApprove: true,
        };
        if (request.takerAddress != null) {
            quoteToCallsRequest.takerAddress = request.takerAddress;
        }
        const result = await quoteToCalls(quoteToCallsRequest, {
            baseUrl: apiBase,
        });
        const calls = normalizeAvnuCalls(result.calls, "AVNU build returned no calls");
        return {
            calls,
            quote: toSwapQuote({
                quote,
                routeCallCount: calls.length,
            }),
        };
    }
    fetchQuoteForRequest(request) {
        const quoteRequest = {
            chainId: request.chainId,
            tokenInAddress: request.tokenIn.address,
            tokenOutAddress: request.tokenOut.address,
            amountInBase: request.amountIn.toBase(),
        };
        if (request.takerAddress != null) {
            quoteRequest.takerAddress = request.takerAddress;
        }
        return this.fetchQuote(quoteRequest);
    }
    async fetchQuote(params) {
        return withAvnuApiBaseFallback({
            apiBasesByChain: this.apiBases,
            chainId: params.chainId,
            feature: "quote",
            action: "quote",
            run: async (apiBase) => {
                const quotesRequest = {
                    sellTokenAddress: params.tokenInAddress,
                    buyTokenAddress: params.tokenOutAddress,
                    sellAmount: params.amountInBase,
                    size: this.quotesPageSize,
                };
                if (params.takerAddress != null) {
                    quotesRequest.takerAddress = params.takerAddress;
                }
                const quotes = await getQuotes(quotesRequest, { baseUrl: apiBase });
                if (!quotes.length) {
                    throw new Error("AVNU quote returned no routes");
                }
                return { quote: quotes[0], apiBase };
            },
            formatFinalError: (failures) => `AVNU quote returned no routes for this pair/amount. Try a larger amount, another token pair, or switch source. (${failures.join(" | ")})`,
        });
    }
}
