// starkzap/dist/src/bridge/ethereum/EtherToken.js
import { Amount, EthereumBridgeToken } from "../../types/index.js";
import { Contract, getAddress, } from "ethers";
import ERC20_ABI from "../../abi/ethereum/erc20.json";
import {} from "../../bridge/ethereum/types.js";
import { fromEthereumAddress } from "../../connect/ethersRuntime.js";
export async function ethereumAddress(contract) {
    const target = contract.target;
    if (typeof target === "string") {
        return fromEthereumAddress(target, { getAddress });
    }
    else {
        const address = await target.getAddress();
        return fromEthereumAddress(address, { getAddress });
    }
}
export function intoEthereumToken(bridgeToken, config) {
    return bridgeToken.id === "eth"
        ? EtherToken.create(config.provider)
        : ERC20EthereumToken.create(bridgeToken.address, config.provider);
}
export class ERC20EthereumToken {
    static create(address, provider) {
        const contract = new Contract(address, ERC20_ABI, provider);
        return new ERC20EthereumToken(contract);
    }
    constructor(contract) {
        this.contract = contract;
    }
    metadata() {
        this._metadata ?? (this._metadata = Promise.all([
            this.contract.getFunction("name")(),
            this.contract.getFunction("symbol")(),
            this.contract.getFunction("decimals")(),
        ]).then(([name, symbol, decimals]) => {
            return {
                name,
                symbol,
                decimals: Number(decimals),
            };
        }));
        return this._metadata;
    }
    async name() {
        return (await this.metadata()).name;
    }
    async symbol() {
        return (await this.metadata()).symbol;
    }
    async decimals() {
        return (await this.metadata()).decimals;
    }
    async balanceOf(account) {
        const balance = await this.contract.getFunction("balanceOf")(account);
        return this.amount(balance);
    }
    async allowance(owner, spender) {
        const allowance = await this.contract.getFunction("allowance")(owner, spender);
        return this.amount(allowance);
    }
    getContract(signer) {
        if (signer) {
            return this.contract.connect(signer);
        }
        else {
            return this.contract;
        }
    }
    async approve(spender, amount, signer) {
        const contract = this.getContract(signer);
        return await contract
            .getFunction("approve")
            .populateTransaction(spender, amount.toBase());
    }
    async amount(amount) {
        const decimals = await this.decimals();
        const symbol = await this.symbol();
        return Amount.fromRaw(amount, decimals, symbol);
    }
    isNativeEth() {
        return false;
    }
    async getAddress() {
        return ethereumAddress(this.getContract());
    }
}
export class EtherToken {
    static create(provider) {
        return new EtherToken(provider);
    }
    constructor(_provider) {
        this._provider = _provider;
    }
    async name() {
        return "Ether";
    }
    async symbol() {
        return "ETH";
    }
    async decimals() {
        return 18;
    }
    async balanceOf(account) {
        const amount = await this._provider.getBalance(account, "pending");
        return this.amount(amount);
    }
    async amount(amount) {
        const decimals = await this.decimals();
        const symbol = await this.symbol();
        return Amount.fromRaw(amount, decimals, symbol);
    }
    async allowance() {
        return null;
    }
    async approve() {
        return null;
    }
    getContract() {
        return null;
    }
    isNativeEth() {
        return true;
    }
}
