// starkzap/dist/src/bridge/ethereum/EthereumBridge.js
import { Amount, EthereumBridgeToken, } from "../../types/index.js";
import { intoEthereumToken, } from "../../bridge/ethereum/EtherToken.js";
import {} from "../../bridge/ethereum/types.js";
import { Contract, getAddress, isError, } from "ethers";
import { FeeErrorCause, TransactionErrorCause } from "../../types/errors.js";
import CANONICAL_BRIDGE_ABI from "../../abi/ethereum/canonicalBridge.json";
import { fromEthereumAddress } from "../../connect/ethersRuntime.js";
export class EthereumBridge {
    constructor(bridgeToken, config, starknetWallet, bridgeAbi = CANONICAL_BRIDGE_ABI) {
        this.bridgeToken = bridgeToken;
        this.config = config;
        this.starknetWallet = starknetWallet;
        this.allowanceCache = {
            current: null,
            timestamp: 0,
        };
        this.token = intoEthereumToken(bridgeToken, config);
        this.bridge = new Contract(bridgeToken.bridgeAddress, bridgeAbi, config.signer);
    }
    async getAvailableDepositBalance(account) {
        return this.token.balanceOf(account);
    }
    async getAllowance() {
        const allowanceSpender = await this.getAllowanceSpender();
        if (!allowanceSpender) {
            return null;
        }
        if (Date.now() - this.allowanceCache.timestamp >
            EthereumBridge.ALLOWANCE_CACHE_TTL) {
            const signerAddress = await this.config.signer.getAddress();
            const allowance = await this.token.allowance(fromEthereumAddress(signerAddress, { getAddress }), allowanceSpender);
            this.setCachedAllowance(allowance);
        }
        return this.allowanceCache.current;
    }
    async approveSpendingOf(amount) {
        const spender = await this.getAllowanceSpender();
        if (!spender) {
            return;
        }
        const allowance = await this.getAllowance();
        if (!allowance) {
            return;
        }
        if (!allowance.lt(amount)) {
            return;
        }
        const tx = await this.token.approve(spender, amount, this.config.signer);
        if (!tx) {
            return;
        }
        const response = await this.execute(tx);
        const receipt = await response.wait();
        if (!receipt?.status) {
            throw new Error(TransactionErrorCause.APPROVE_FAILED);
        }
        await this.updateAllowanceFromReceipt(receipt);
    }
    async execute(tx) {
        try {
            return (await this.config.signer.sendTransaction(tx));
        }
        catch (e) {
            if (isError(e, "ACTION_REJECTED")) {
                throw new Error(TransactionErrorCause.USER_REJECTED);
            }
            if (isError(e, "INSUFFICIENT_FUNDS")) {
                throw new Error(TransactionErrorCause.INSUFFICIENT_BALANCE);
            }
            // TODO be more specific with other ethers errors
            throw e;
        }
    }
    async populateTransaction(details) {
        return await this.bridge
            .getFunction(details.method)
            .populateTransaction(...details.args, details.transaction);
    }
    ethAmount(value) {
        return Amount.fromRaw(value, 18, "ETH");
    }
    async estimateApprovalFee() {
        if (this.token.isNativeEth()) {
            return { approvalFee: this.ethAmount(0n) };
        }
        const spender = await this.getAllowanceSpender();
        if (!spender) {
            return { approvalFee: this.ethAmount(0n) };
        }
        const contract = this.token.getContract();
        if (!contract) {
            return {
                approvalFee: this.ethAmount(0n),
                approvalFeeError: FeeErrorCause.NO_TOKEN_CONTRACT,
            };
        }
        try {
            const approvalTransaction = await this.token.approve(spender, await this.token.amount(1n), this.config.signer);
            if (!approvalTransaction) {
                return {
                    approvalFee: this.ethAmount(0n),
                    approvalFeeError: FeeErrorCause.NO_TOKEN_CONTRACT,
                };
            }
            const from = await this.config.signer.getAddress();
            const [approvalGasRequirement, gasPrice] = await Promise.all([
                this.config.provider.estimateGas({ ...approvalTransaction, from }),
                this.getEthereumGasPrice(),
            ]);
            const approvalFee = approvalGasRequirement * gasPrice;
            return { approvalFee: this.ethAmount(approvalFee) };
        }
        catch {
            return {
                approvalFee: this.ethAmount(0n),
                approvalFeeError: FeeErrorCause.APPROVAL_FEE_ERROR,
            };
        }
    }
    clearCachedAllowance() {
        this.allowanceCache.timestamp = -1;
    }
    setCachedAllowance(newValue) {
        this.allowanceCache = {
            current: newValue,
            timestamp: Date.now(),
        };
    }
    async updateAllowanceFromReceipt(receipt) {
        const tokenInterface = this.token.getContract()?.interface;
        if (!tokenInterface || !receipt.logs)
            return;
        let newAllowance = null;
        for (const log of receipt.logs) {
            let parsedLog;
            try {
                parsedLog = tokenInterface.parseLog(log);
            }
            catch {
                continue;
            }
            if (parsedLog?.name === "Approval" &&
                typeof parsedLog.args?.value === "bigint") {
                newAllowance = parsedLog.args.value;
                break;
            }
        }
        if (newAllowance !== null) {
            const amount = await this.token.amount(newAllowance);
            this.setCachedAllowance(amount);
        }
        else {
            this.clearCachedAllowance();
        }
    }
}
EthereumBridge.ALLOWANCE_CACHE_TTL = 60000;
EthereumBridge.GAS_LIMIT_SAFE_MULTIPLIER = 1.5;
