// starkzap/dist/src/dca/avnu.js
import { cancelDcaToCalls, createDcaToCalls, getDcaOrders, DcaOrderStatus as AvnuDcaOrderStatus, } from "@avnu/avnu-sdk";
import { assertAmountMatchesToken, fromAddress } from "../types/index.js";
import { validateDcaCreateAmounts } from "../dca/utils.js";
import { DEFAULT_AVNU_API_BASES, normalizeAvnuCalls, supportsAvnuChain, withAvnuApiBaseFallback, } from "../utils/avnu.js";
function toHexAmount(value) {
    return `0x${value.toString(16)}`;
}
const DCA_STATUS_TO_AVNU = {
    INDEXING: AvnuDcaOrderStatus.INDEXING,
    ACTIVE: AvnuDcaOrderStatus.ACTIVE,
    CLOSED: AvnuDcaOrderStatus.CLOSED,
};
function toPricingStrategy(strategy) {
    if (!strategy) {
        return {};
    }
    const minBuyAmount = strategy.minBuyAmount;
    const maxBuyAmount = strategy.maxBuyAmount;
    const minBuyAmountBase = minBuyAmount?.toBase();
    const maxBuyAmountBase = maxBuyAmount?.toBase();
    if (minBuyAmountBase != null &&
        maxBuyAmountBase != null &&
        minBuyAmountBase > maxBuyAmountBase) {
        throw new Error("DCA pricingStrategy.minBuyAmount cannot exceed pricingStrategy.maxBuyAmount");
    }
    if (minBuyAmountBase == null && maxBuyAmountBase == null) {
        return {};
    }
    return {
        ...(minBuyAmountBase != null && {
            tokenToMinAmount: toHexAmount(minBuyAmountBase),
        }),
        ...(maxBuyAmountBase != null && {
            tokenToMaxAmount: toHexAmount(maxBuyAmountBase),
        }),
    };
}
function validateCreateRequest(request) {
    validateDcaCreateAmounts(request);
    const minBuyAmount = request.pricingStrategy?.minBuyAmount;
    const maxBuyAmount = request.pricingStrategy?.maxBuyAmount;
    if (minBuyAmount) {
        assertAmountMatchesToken(minBuyAmount, request.buyToken);
    }
    if (maxBuyAmount) {
        assertAmountMatchesToken(maxBuyAmount, request.buyToken);
    }
}
function mapPricingStrategy(strategy) {
    const pricingStrategy = {};
    if ("tokenToMinAmount" in strategy && strategy.tokenToMinAmount) {
        pricingStrategy.minBuyAmountBase = BigInt(strategy.tokenToMinAmount);
    }
    if ("tokenToMaxAmount" in strategy && strategy.tokenToMaxAmount) {
        pricingStrategy.maxBuyAmountBase = BigInt(strategy.tokenToMaxAmount);
    }
    return pricingStrategy;
}
function mapTrade(trade) {
    const mappedTrade = {
        sellAmountBase: trade.sellAmount,
        expectedTradeDate: trade.expectedTradeDate,
        status: trade.status,
    };
    if (trade.sellAmountInUsd != null) {
        mappedTrade.sellAmountInUsd = trade.sellAmountInUsd;
    }
    if (trade.buyAmount != null) {
        mappedTrade.buyAmountBase = trade.buyAmount;
    }
    if (trade.buyAmountInUsd != null) {
        mappedTrade.buyAmountInUsd = trade.buyAmountInUsd;
    }
    if (trade.actualTradeDate) {
        mappedTrade.actualTradeDate = trade.actualTradeDate;
    }
    if (trade.txHash) {
        mappedTrade.txHash = trade.txHash;
    }
    if (trade.errorReason) {
        mappedTrade.errorReason = trade.errorReason;
    }
    return mappedTrade;
}
function mapOrder(order) {
    const mappedOrder = {
        id: order.id,
        providerId: "avnu",
        blockNumber: order.blockNumber,
        timestamp: order.timestamp,
        traderAddress: fromAddress(order.traderAddress),
        orderAddress: fromAddress(order.orderAddress),
        creationTransactionHash: order.creationTransactionHash,
        orderClassHash: order.orderClassHash,
        sellTokenAddress: fromAddress(order.sellTokenAddress),
        sellAmountBase: order.sellAmount,
        sellAmountPerCycleBase: order.sellAmountPerCycle,
        buyTokenAddress: fromAddress(order.buyTokenAddress),
        startDate: order.startDate,
        endDate: order.endDate,
        frequency: order.frequency,
        iterations: order.iterations,
        status: order.status,
        pricingStrategy: mapPricingStrategy(order.pricingStrategy),
        amountSoldBase: order.amountSold,
        amountBoughtBase: order.amountBought,
        averageAmountBoughtBase: order.averageAmountBought,
        executedTradesCount: order.executedTradesCount,
        cancelledTradesCount: order.cancelledTradesCount,
        pendingTradesCount: order.pendingTradesCount,
        trades: order.trades.map(mapTrade),
    };
    if (order.closeDate) {
        mappedOrder.closeDate = order.closeDate;
    }
    return mappedOrder;
}
export class AvnuDcaProvider {
    constructor(options = {}) {
        this.id = "avnu";
        this.apiBases = {
            SN_MAIN: options.apiBases?.SN_MAIN ?? [...DEFAULT_AVNU_API_BASES.SN_MAIN],
            SN_SEPOLIA: options.apiBases?.SN_SEPOLIA ?? [
                ...DEFAULT_AVNU_API_BASES.SN_SEPOLIA,
            ],
        };
    }
    supportsChain(chainId) {
        return supportsAvnuChain(chainId);
    }
    async getOrders(context, request) {
        const avnuRequest = {
            traderAddress: request.traderAddress,
        };
        if (request.status) {
            avnuRequest.status = DCA_STATUS_TO_AVNU[request.status];
        }
        if (request.page != null) {
            avnuRequest.page = request.page;
        }
        if (request.size != null) {
            avnuRequest.size = request.size;
        }
        if (request.sort) {
            avnuRequest.sort = request.sort;
        }
        const page = await withAvnuApiBaseFallback({
            apiBasesByChain: this.apiBases,
            chainId: context.chainId,
            feature: "DCA",
            action: "get DCA orders",
            run: (baseUrl) => getDcaOrders(avnuRequest, { baseUrl }),
        });
        return {
            content: page.content.map(mapOrder),
            totalPages: page.totalPages,
            totalElements: page.totalElements,
            size: page.size,
            pageNumber: page.number,
        };
    }
    async prepareCreate(context, request) {
        validateCreateRequest(request);
        const createRequest = {
            sellTokenAddress: request.sellToken.address,
            buyTokenAddress: request.buyToken.address,
            sellAmount: toHexAmount(request.sellAmount.toBase()),
            sellAmountPerCycle: toHexAmount(request.sellAmountPerCycle.toBase()),
            frequency: {
                toJSON: () => request.frequency,
                toISOString: () => request.frequency,
            },
            pricingStrategy: toPricingStrategy(request.pricingStrategy),
            traderAddress: request.traderAddress,
        };
        const calls = await withAvnuApiBaseFallback({
            apiBasesByChain: this.apiBases,
            chainId: context.chainId,
            feature: "DCA",
            action: "prepare DCA create",
            run: async (baseUrl) => {
                const response = await createDcaToCalls(createRequest, { baseUrl });
                return normalizeAvnuCalls(response.calls, "AVNU DCA create returned no calls");
            },
        });
        return {
            providerId: this.id,
            action: "create",
            calls,
        };
    }
    async prepareCancel(context, request) {
        if (!request.orderAddress) {
            throw new Error("AVNU DCA cancel requires an orderAddress");
        }
        const orderAddress = request.orderAddress;
        const calls = await withAvnuApiBaseFallback({
            apiBasesByChain: this.apiBases,
            chainId: context.chainId,
            feature: "DCA",
            action: "prepare DCA cancel",
            run: async (baseUrl) => {
                const response = await cancelDcaToCalls(orderAddress, { baseUrl });
                return normalizeAvnuCalls(response.calls, "AVNU DCA cancel returned no calls");
            },
        });
        return {
            providerId: this.id,
            action: "cancel",
            calls,
            orderAddress,
        };
    }
}
