// starkzap/dist/src/dca/client.js
import { resolveWalletAddress } from "../types/index.js";
import { AvnuDcaProvider } from "../dca/avnu.js";
import { assertDcaContext, hydrateDcaCancelInput, hydrateDcaCreateInput, hydrateDcaOrdersInput, resolveDcaSource, } from "../dca/utils.js";
import { resolveSwapInput } from "../swap/utils.js";
export class DcaClient {
    constructor(context, defaultProvider) {
        this.defaultProviderId = null;
        this.context = context;
        this.providers = new Map();
        this.registerProvider(defaultProvider ?? new AvnuDcaProvider(), true);
    }
    registerProvider(provider, makeDefault = false) {
        this.providers.set(provider.id, provider);
        if (makeDefault || this.defaultProviderId == null) {
            this.defaultProviderId = provider.id;
        }
    }
    setDefaultProvider(providerId) {
        this.getDcaProvider(providerId);
        this.defaultProviderId = providerId;
    }
    getDcaProvider(providerId) {
        const provider = this.providers.get(providerId);
        if (!provider) {
            throw new Error(`Unknown DCA provider "${providerId}". Registered providers: ${this.listProviders().join(", ")}`);
        }
        return provider;
    }
    getDefaultDcaProvider() {
        if (!this.defaultProviderId) {
            throw new Error("No default DCA provider configured");
        }
        return this.getDcaProvider(this.defaultProviderId);
    }
    listProviders() {
        return Array.from(this.providers.keys());
    }
    async getOrders(request = {}) {
        const provider = this.resolveRequestProvider(request.provider);
        return provider.getOrders(this.providerContext(), hydrateDcaOrdersInput(request, this.context.address));
    }
    async prepareCreate(request) {
        const provider = this.resolveRequestProvider(request.provider);
        const prepared = await provider.prepareCreate(this.providerContext(), hydrateDcaCreateInput(request, this.context.address));
        this.assertPreparedCalls(prepared, provider.id);
        return prepared;
    }
    async create(request, options) {
        const prepared = await this.prepareCreate(request);
        return this.context.execute(prepared.calls, options);
    }
    async prepareCancel(request) {
        const provider = this.resolveRequestProvider(request.provider);
        const prepared = await provider.prepareCancel(this.providerContext(), hydrateDcaCancelInput(request));
        this.assertPreparedCalls(prepared, provider.id);
        return prepared;
    }
    async cancel(request, options) {
        const prepared = await this.prepareCancel(request);
        return this.context.execute(prepared.calls, options);
    }
    async previewCycle(request) {
        const takerAddress = request.traderAddress != null
            ? resolveWalletAddress(request.traderAddress)
            : undefined;
        const { provider, request: resolvedRequest } = resolveSwapInput({
            tokenIn: request.sellToken,
            tokenOut: request.buyToken,
            amountIn: request.sellAmountPerCycle,
            ...(request.swapProvider != null && { provider: request.swapProvider }),
            ...(request.chainId != null && { chainId: request.chainId }),
            ...(takerAddress != null && { takerAddress }),
            ...(request.slippageBps != null && {
                slippageBps: request.slippageBps,
            }),
        }, {
            walletChainId: this.context.getChainId(),
            takerAddress: this.context.address,
            providerResolver: this.context,
        });
        return provider.getQuote(resolvedRequest);
    }
    resolveRequestProvider(source) {
        const provider = resolveDcaSource(source, this);
        assertDcaContext(provider, this.context.getChainId());
        return provider;
    }
    providerContext() {
        return {
            chainId: this.context.getChainId(),
            rpcProvider: this.context.getProvider(),
            walletAddress: this.context.address,
        };
    }
    assertPreparedCalls(prepared, providerId) {
        if (prepared.calls.length > 0) {
            return;
        }
        throw new Error(`DCA provider "${providerId}" returned no calls`);
    }
}
