// starkzap/dist/src/bridge/ethereum/cctp/CCTPFees.js
import { ETH_FAST_TRANSFER_FEE_BP, ETHEREUM_DOMAIN_ID, getFinalityThreshold, LIVE_DOMAIN, SANDBOX_DOMAIN, STARKNET_DOMAIN_ID, STARKNET_FAST_TRANSFER_FEE_BP, } from "../../../bridge/ethereum/cctp/constants.js";
export var BridgeDirection;
(function (BridgeDirection) {
    BridgeDirection[BridgeDirection["WITHDRAW_FROM_STARKNET"] = 0] = "WITHDRAW_FROM_STARKNET";
    BridgeDirection[BridgeDirection["DEPOSIT_TO_STARKNET"] = 1] = "DEPOSIT_TO_STARKNET";
})(BridgeDirection || (BridgeDirection = {}));
export class CCTPFees {
    constructor() { }
    static getInstance() {
        if (!CCTPFees.instance) {
            CCTPFees.instance = new CCTPFees();
        }
        return CCTPFees.instance;
    }
    async getMinimumFeeBps(direction, chainId, fastTransfer) {
        try {
            const feeData = await this.getFees(direction, chainId);
            if (!Array.isArray(feeData)) {
                return this.getFallbackFee(direction, fastTransfer);
            }
            const targetThreshold = getFinalityThreshold(fastTransfer);
            const fee = feeData.find((f) => f.finalityThreshold === targetThreshold);
            if (fee) {
                return fee.minimumFee;
            }
            return this.getFallbackFee(direction, fastTransfer);
        }
        catch (error) {
            console.error("Failed to get transfer fee, using fallback:", error);
            return this.getFallbackFee(direction, fastTransfer);
        }
    }
    async getFees(direction, chainId) {
        // Possibly cache these fees.
        return this.fetchFees(direction, chainId);
    }
    async fetchFees(direction, chainId) {
        const source = direction === BridgeDirection.DEPOSIT_TO_STARKNET
            ? ETHEREUM_DOMAIN_ID
            : STARKNET_DOMAIN_ID;
        const destination = direction === BridgeDirection.DEPOSIT_TO_STARKNET
            ? STARKNET_DOMAIN_ID
            : ETHEREUM_DOMAIN_ID;
        let domainUrl;
        if (chainId.isMainnet()) {
            domainUrl = LIVE_DOMAIN;
        }
        else {
            domainUrl = SANDBOX_DOMAIN;
        }
        const url = `${domainUrl}/v2/burn/USDC/fees/${source}/${destination}`;
        const response = await fetch(url);
        if (!response.ok) {
            throw new Error(`Failed to fetch fees from Circle API: ${response.statusText}`);
        }
        const data = (await response.json());
        return data.data;
    }
    getFallbackFee(direction, fastTransfer) {
        if (!fastTransfer) {
            return 0;
        }
        return direction === BridgeDirection.DEPOSIT_TO_STARKNET
            ? ETH_FAST_TRANSFER_FEE_BP
            : STARKNET_FAST_TRANSFER_FEE_BP;
    }
}
