// starkzap/dist/src/bridge/ethereum/cctp/CCTPBridge.js
import { Amount, fromAddress, } from "../../../types/index.js";
import { ERC20EthereumToken } from "../../../bridge/ethereum/EtherToken.js";
import { getAddress, Interface } from "ethers";
import { FeeErrorCause } from "../../../types/errors.js";
import { BridgeDirection, CCTPFees } from "../../../bridge/ethereum/cctp/CCTPFees.js";
import { getFinalityThreshold, STARKNET_DOMAIN_ID, } from "../../../bridge/ethereum/cctp/constants.js";
import { EthereumBridge } from "../../../bridge/ethereum/EthereumBridge.js";
import { fromEthereumAddress } from "../../../connect/ethersRuntime.js";
export class CCTPBridge extends EthereumBridge {
    constructor() {
        super(...arguments);
        this.cctpFees = CCTPFees.getInstance();
    }
    async deposit(recipient, amount, options) {
        await this.approveSpendingOf(amount);
        const txRequest = await this.createDepositForBurnTransaction(recipient, amount, undefined, options?.fastTransfer);
        const txResponse = await this.execute(txRequest);
        this.clearCachedAllowance();
        return { hash: txResponse.hash };
    }
    async getDepositFeeEstimate(options) {
        const fastTransfer = options?.fastTransfer;
        const minimalAmount = this.usdcAmount(2n);
        const [allowance, approvalFeeData, feeData, minimumFeeBps] = await Promise.all([
            this.getAllowance(),
            this.estimateApprovalFee(),
            this.config.provider.getFeeData(),
            this.cctpFees.getMinimumFeeBps(BridgeDirection.DEPOSIT_TO_STARKNET, this.starknetWallet.getChainId(), fastTransfer),
        ]);
        const gasPrice = feeData.maxFeePerGas ?? feeData.gasPrice ?? 0n;
        const defaultL1Fee = this.ethAmount(CCTPBridge.DEFAULT_CCTP_DEPOSIT_GAS * gasPrice);
        if (!allowance || allowance.lt(minimalAmount)) {
            return {
                l1Fee: defaultL1Fee,
                l2Fee: CCTPBridge.ZERO_ETH,
                fastTransferBpFee: minimumFeeBps,
                ...approvalFeeData,
            };
        }
        else {
            const txRequest = await this.createDepositForBurnTransaction(CCTPBridge.DUMMY_SN_ADDRESS, minimalAmount, minimumFeeBps, fastTransfer);
            try {
                const gasEstimate = await this.config.signer.estimateGas(txRequest);
                const l1Fee = gasEstimate * gasPrice;
                return {
                    l1Fee: this.ethAmount(l1Fee),
                    l2Fee: CCTPBridge.ZERO_ETH,
                    fastTransferBpFee: minimumFeeBps,
                    ...approvalFeeData,
                };
            }
            catch {
                return {
                    l1Fee: defaultL1Fee,
                    l1FeeError: FeeErrorCause.GENERIC_L1_FEE_ERROR,
                    l2Fee: CCTPBridge.ZERO_ETH,
                    fastTransferBpFee: minimumFeeBps,
                    ...approvalFeeData,
                };
            }
        }
    }
    getAllowanceSpender() {
        if (this.starknetWallet.getChainId().isMainnet()) {
            return Promise.resolve(CCTPBridge.MAINNET_TOKEN_MESSENGER);
        }
        else {
            return Promise.resolve(CCTPBridge.SEPOLIA_TOKEN_MESSENGER);
        }
    }
    async getEthereumGasPrice() {
        const feeData = await this.config.provider.getFeeData();
        return feeData.maxFeePerGas ?? feeData.gasPrice ?? 0n;
    }
    usdcAmount(value) {
        return Amount.fromRaw(value, 6, "USDC");
    }
    async createDepositForBurnTransaction(recipient, amount, fastTransferFeeBps, fastTransfer) {
        const usdcToken = this.token;
        const usdcAddress = await usdcToken.getAddress();
        const feeBps = fastTransferFeeBps ??
            (await this.cctpFees.getMinimumFeeBps(BridgeDirection.DEPOSIT_TO_STARKNET, this.starknetWallet.getChainId(), fastTransfer));
        const maxFee = this.calculateMaxFee(amount, feeBps);
        const calldata = CCTPBridge.TOKEN_MESSENGER_INTERFACE.encodeFunctionData("depositForBurn", [
            amount.toBase(),
            STARKNET_DOMAIN_ID,
            recipient,
            usdcAddress,
            "0x0000000000000000000000000000000000000000000000000000000000000000",
            maxFee.toBase(),
            getFinalityThreshold(fastTransfer),
        ]);
        return {
            to: await this.getAllowanceSpender(),
            data: calldata,
        };
    }
    calculateMaxFee(amount, feeBasisPoints) {
        const numerator = amount.toBase() * BigInt(feeBasisPoints);
        const divisor = 10000n; // Basis points
        // Round up by adding (divisor - 1) before dividing
        const result = (numerator + divisor - 1n) / divisor;
        return this.usdcAmount(result);
    }
}
CCTPBridge.MAINNET_TOKEN_MESSENGER = fromEthereumAddress("0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d", { getAddress });
CCTPBridge.SEPOLIA_TOKEN_MESSENGER = fromEthereumAddress("0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA", { getAddress });
CCTPBridge.DEFAULT_CCTP_DEPOSIT_GAS = 104581n;
CCTPBridge.TOKEN_MESSENGER_INTERFACE = new Interface([
    "function depositForBurn(uint256 amount, uint32 destinationDomain, bytes32 mintRecipient, address burnToken, bytes32 destinationCaller, uint256 maxFee, uint32 minFinalityThreshold)",
]);
CCTPBridge.DUMMY_SN_ADDRESS = fromAddress("0x0000000000000000000000000000000000000000000000000000000000000001");
CCTPBridge.ZERO_ETH = Amount.fromRaw(0n, 18, "ETH");
