// starkzap/dist/src/utils/avnu.js
import { CallData } from "starknet";
import { BASE_URL, SEPOLIA_BASE_URL } from "@avnu/avnu-sdk";
export const DEFAULT_AVNU_API_BASES = {
    SN_MAIN: [BASE_URL],
    SN_SEPOLIA: [SEPOLIA_BASE_URL, BASE_URL],
};
export function supportsAvnuChain(chainId) {
    const literal = chainId.toLiteral();
    return literal === "SN_MAIN" || literal === "SN_SEPOLIA";
}
export function getAvnuApiBases(apiBasesByChain, chainId, feature) {
    const literal = chainId.toLiteral();
    let apiBases;
    if (literal === "SN_MAIN") {
        apiBases = apiBasesByChain.SN_MAIN;
    }
    else if (literal === "SN_SEPOLIA") {
        apiBases = apiBasesByChain.SN_SEPOLIA;
    }
    else {
        throw new Error(`Unsupported chain for AVNU ${feature}: ${literal}`);
    }
    if (apiBases.length === 0) {
        throw new Error(`No AVNU API base configured for chain: ${literal}`);
    }
    return [...apiBases];
}
export function describeAvnuError(error) {
    if (error instanceof Error) {
        return error.message;
    }
    return String(error);
}
export async function withAvnuApiBaseFallback(params) {
    const failures = [];
    for (const apiBase of getAvnuApiBases(params.apiBasesByChain, params.chainId, params.feature)) {
        try {
            return await params.run(apiBase);
        }
        catch (error) {
            failures.push(`${apiBase}: ${describeAvnuError(error)}`);
        }
    }
    throw new Error(params.formatFinalError?.(failures) ??
        `AVNU ${params.action} failed (${failures.join(" | ")})`);
}
export function normalizeAvnuCalls(calls, emptyMessage) {
    if (calls.length === 0) {
        throw new Error(emptyMessage);
    }
    return calls.map((call) => ({
        contractAddress: call.contractAddress,
        entrypoint: `${call.entrypoint}`,
        calldata: CallData.compile(call.calldata ?? []),
    }));
}
