// @fatsolutions/she/lib/exceptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LengthMismatchError = exports.PreconditionError = exports.OutOfRangeError = exports.InvalidInputError = exports.ProtocolError = void 0;
/**
 * Base class for all protocol-related errors.
 */
class ProtocolError extends Error {
    errorCode;
    constructor(message, errorCode) {
        super(message);
        this.errorCode = errorCode;
        this.name = this.constructor.name;
        // Maintains proper stack trace for where our error was thrown (only available on V8)
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, this.constructor);
        }
    }
}
exports.ProtocolError = ProtocolError;
/**
 * Thrown when input parameters have invalid lengths, formats, or values.
 */
class InvalidInputError extends ProtocolError {
    constructor(message) {
        super(message, 'INVALID_INPUT');
    }
}
exports.InvalidInputError = InvalidInputError;
/**
 * Thrown when input values are out of valid ranges.
 */
class OutOfRangeError extends ProtocolError {
    value;
    range;
    constructor(message, value, range) {
        super(message, 'OUT_OF_RANGE');
        this.value = value;
        this.range = range;
    }
}
exports.OutOfRangeError = OutOfRangeError;
/**
 * Thrown when preconditions for a protocol operation are not met.
 */
class PreconditionError extends ProtocolError {
    constructor(message) {
        super(message, 'PRECONDITION_FAILED');
    }
}
exports.PreconditionError = PreconditionError;
/**
 * Thrown when array lengths don't match expected requirements.
 */
class LengthMismatchError extends InvalidInputError {
    constructor(expected, actual, context = 'arrays') {
        super(`Length mismatch: expected ${expected} but got ${actual} for ${context}`);
    }
}
exports.LengthMismatchError = LengthMismatchError;
