// starkzap/dist/src/bridge/solana/registry.js
function buildTestnetChainMap(hyperlane) {
    return {
        solanatestnet: {
            ...hyperlane.registry.solanatestnet,
            mailbox: hyperlane.registry.solanatestnetAddresses.mailbox,
        },
        starknetsepolia: {
            ...hyperlane.registry.starknetsepolia,
            mailbox: hyperlane.registry.starknetsepoliaAddresses.mailbox,
        },
    };
}
function buildMainnetChainMap(hyperlane) {
    return {
        solanamainnet: {
            ...hyperlane.registry.solanamainnet,
            mailbox: hyperlane.registry.solanamainnetAddresses.mailbox,
        },
        starknet: {
            ...hyperlane.registry.starknet,
            mailbox: hyperlane.registry.starknetAddresses.mailbox,
        },
    };
}
const STARKNET_CHAIN_TO_HYPERLANE = {
    SN_MAIN: { starknet: "starknet", solana: "solanamainnet" },
    SN_SEPOLIA: { starknet: "starknetsepolia", solana: "solanatestnet" },
};
export function hyperlaneChainName(chainId, hyperlaneChain) {
    const hyperlaneConfig = STARKNET_CHAIN_TO_HYPERLANE[chainId.toLiteral()];
    if (!hyperlaneConfig) {
        throw new Error(`Unknown starknet chain ID: ${chainId.toLiteral()}`);
    }
    const name = hyperlaneConfig[hyperlaneChain];
    if (!name) {
        throw new Error(`Unknown chain "${hyperlaneChain}" for network ${chainId.toLiteral()}`);
    }
    return name;
}
export function setupMultiProtocolProvider(config, starknetWallet, hyperlane) {
    const chainId = starknetWallet.getChainId();
    const chains = chainId.isMainnet()
        ? buildMainnetChainMap(hyperlane)
        : buildTestnetChainMap(hyperlane);
    const MultiProtocolProviderCtor = hyperlane.sdk.MultiProtocolProvider;
    const ProviderType = hyperlane.sdk.ProviderType;
    const multiProvider = new MultiProtocolProviderCtor(chains);
    const solanaProvider = {
        type: ProviderType.SolanaWeb3,
        // `connection` is intentionally opaque in public types to avoid exporting
        // @solana/web3.js symbols in SDK declarations.
        provider: config.connection,
    };
    multiProvider.setProvider(hyperlaneChainName(chainId, "solana"), solanaProvider);
    return multiProvider;
}
export function bridgeTokenToHyperlaneToken(token, chainId, hyperlaneChain, hyperlane) {
    const isStarknet = hyperlaneChain === "starknet";
    const bridgeAddress = isStarknet ? token.starknetBridge : token.bridgeAddress;
    const collateralAddress = isStarknet ? token.starknetAddress : token.address;
    const isNative = token.id === "sol";
    const TokenStandard = hyperlane.sdk.TokenStandard;
    const tokenStandard = hyperlaneChain === "starknet"
        ? TokenStandard.StarknetHypSynthetic
        : isNative
            ? TokenStandard.SealevelHypNative
            : TokenStandard.SealevelHypCollateral;
    const HyperlaneTokenCtor = hyperlane.sdk.Token;
    return new HyperlaneTokenCtor({
        symbol: token.symbol,
        name: token.name,
        decimals: token.decimals,
        chainName: hyperlaneChainName(chainId, hyperlaneChain),
        addressOrDenom: bridgeAddress,
        collateralAddressOrDenom: collateralAddress,
        standard: tokenStandard,
    });
}
