// starkzap/dist/src/bridge/solana/SolanaHyperlaneBridge.js
import { Amount, } from "../../types/index.js";
import { FeeErrorCause } from "../../types/errors.js";
import { bridgeTokenToHyperlaneToken, hyperlaneChainName, setupMultiProtocolProvider, } from "../../bridge/solana/registry.js";
import { loadHyperlane, } from "../../bridge/solana/hyperlaneRuntime.js";
// https://github.com/hyperlane-xyz/hyperlane-warp-ui-template/blob/21ac2754c69f69d056a39bcc664531d6118fee0c/src/consts/chains.ts#L68
const SOLANA_RENT_ESTIMATE = BigInt(Math.round(0.00411336 * 1e9));
export class SolanaHyperlaneBridge {
    constructor(bridgeToken, config, starknetWallet, hyperlane, multiProvider, warpCore, solanaToken, starknetChain, solanaChain) {
        this.bridgeToken = bridgeToken;
        this.config = config;
        this.starknetWallet = starknetWallet;
        this.hyperlane = hyperlane;
        this.multiProvider = multiProvider;
        this.warpCore = warpCore;
        this.solanaToken = solanaToken;
        this.starknetChain = starknetChain;
        this.solanaChain = solanaChain;
    }
    static async create(bridgeToken, config, starknetWallet) {
        const hyperlane = await loadHyperlane("Solana bridge operations");
        const chainId = starknetWallet.getChainId();
        const multiProvider = setupMultiProtocolProvider(config, starknetWallet, hyperlane);
        const solanaToken = bridgeTokenToHyperlaneToken(bridgeToken, chainId, "solana", hyperlane);
        const starknetToken = bridgeTokenToHyperlaneToken(bridgeToken, chainId, "starknet", hyperlane);
        const WarpCoreCtor = hyperlane.sdk.WarpCore;
        const warpCore = new WarpCoreCtor(multiProvider, [
            solanaToken,
            starknetToken,
        ]);
        return new SolanaHyperlaneBridge(bridgeToken, config, starknetWallet, hyperlane, multiProvider, warpCore, solanaToken, hyperlaneChainName(chainId, "starknet"), hyperlaneChainName(chainId, "solana"));
    }
    async deposit(recipient, amount) {
        const TokenAmountCtor = this.hyperlane.sdk.TokenAmount;
        const transactions = (await this.warpCore.getTransferRemoteTxs({
            destination: this.starknetChain,
            originTokenAmount: new TokenAmountCtor(amount.toBase(), this.solanaToken),
            sender: this.config.address,
            recipient,
        }));
        if (transactions.length === 0) {
            throw new Error("Hyperlane returned no deposit transactions.");
        }
        let lastSignature = "";
        for (const tx of transactions) {
            lastSignature = await this.config.provider.signAndSendTransaction(tx.transaction);
        }
        return { hash: lastSignature };
    }
    async getDepositFeeEstimate() {
        const interchainResult = await this.estimateDepositInterchainFee();
        const localResult = await this.estimateDepositLocalFee(interchainResult.interchainFee);
        const estimate = {
            localFee: this.solAmount(localResult.localFee.amount),
            interchainFee: this.solAmount(interchainResult.interchainFee.amount),
        };
        if (localResult.localFeeError) {
            estimate.localFeeError = localResult.localFeeError;
        }
        if (interchainResult.interchainFeeError) {
            estimate.interchainFeeError = interchainResult.interchainFeeError;
        }
        return estimate;
    }
    async getAvailableDepositBalance(account) {
        const balance = await this.solanaToken.getBalance(this.multiProvider, account);
        const raw = balance?.amount ?? 0n;
        return Amount.fromRaw(raw, this.bridgeToken.decimals, this.bridgeToken.symbol);
    }
    async getAllowance() {
        return null;
    }
    async estimateDepositInterchainFee() {
        try {
            const quote = await this.warpCore.getInterchainTransferFee({
                destination: this.starknetChain,
                originToken: this.solanaToken,
                sender: this.config.address,
            });
            return { interchainFee: quote.plus(SOLANA_RENT_ESTIMATE) };
        }
        catch {
            const HyperlaneTokenCtor = this.hyperlane.sdk.Token;
            const TokenAmountCtor = this.hyperlane.sdk.TokenAmount;
            const TokenStandard = this.hyperlane.sdk.TokenStandard;
            const zeroToken = new HyperlaneTokenCtor({
                symbol: "SOL",
                name: "Solana",
                decimals: 9,
                chainName: this.solanaChain,
                addressOrDenom: "native",
                standard: TokenStandard.SealevelHypNative,
            });
            return {
                interchainFee: new TokenAmountCtor(0n, zeroToken),
                interchainFeeError: FeeErrorCause.GENERIC_L1_FEE_ERROR,
            };
        }
    }
    async estimateDepositLocalFee(interchainFee) {
        try {
            const { fee } = await this.warpCore.getLocalTransferFee({
                destination: this.starknetChain,
                originToken: this.solanaToken,
                sender: this.config.address,
                interchainFee,
            });
            const TokenAmountCtor = this.hyperlane.sdk.TokenAmount;
            return {
                localFee: new TokenAmountCtor(BigInt(fee), this.solanaToken),
            };
        }
        catch {
            const TokenAmountCtor = this.hyperlane.sdk.TokenAmount;
            return {
                localFee: new TokenAmountCtor(0n, interchainFee.token),
                localFeeError: FeeErrorCause.GENERIC_L1_FEE_ERROR,
            };
        }
    }
    solAmount(amount) {
        return Amount.fromRaw(amount, 9, "SOL");
    }
}
